<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2019_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg">
<meta name="author" content="Dana Silverbush">
<meta name="author" content="Kirk Gosik">


<meta name="date" content="2019-02-21">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  

<link rel="next" href="scrna-seq-experimental-design.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks"><i class="fa fa-check"></i><b>1.5.1</b> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 09:30 to 17:30</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scrna-seq-experimental-design.html"><a href="scrna-seq-experimental-design.html"><i class="fa fa-check"></i><b>2</b> scRNA-Seq Experimental Design</a></li>
<li class="chapter" data-level="3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>3</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>3.1</b> Common Linux Commands</a></li>
<li class="chapter" data-level="3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#raw-file-formats"><i class="fa fa-check"></i><b>3.2</b> Raw file formats</a></li>
<li class="chapter" data-level="3.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>3.3</b> GEO</a></li>
<li class="chapter" data-level="3.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>3.4</b> Single Cell Portal</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Data Preprocessing</a></li>
<li class="chapter" data-level="5" data-path="processing-scrna-seq-data.html"><a href="processing-scrna-seq-data.html"><i class="fa fa-check"></i><b>5</b> Processing scRNA-Seq data</a></li>
<li class="chapter" data-level="6" data-path="transcriptome-quantification.html"><a href="transcriptome-quantification.html"><i class="fa fa-check"></i><b>6</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>7</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>7.1</b> Installing packages</a><ul>
<li class="chapter" data-level="7.1.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>7.1.1</b> CRAN</a></li>
<li class="chapter" data-level="7.1.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>7.1.2</b> Github</a></li>
<li class="chapter" data-level="7.1.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>7.1.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.1.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>7.1.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>7.2</b> Installation instructions:</a></li>
<li class="chapter" data-level="7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-typesclasses"><i class="fa fa-check"></i><b>7.3</b> Data-types/classes</a><ul>
<li class="chapter" data-level="7.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#numeric"><i class="fa fa-check"></i><b>7.3.1</b> Numeric</a></li>
<li class="chapter" data-level="7.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#characterstring"><i class="fa fa-check"></i><b>7.3.2</b> Character/String</a></li>
<li class="chapter" data-level="7.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#logical"><i class="fa fa-check"></i><b>7.3.3</b> Logical</a></li>
<li class="chapter" data-level="7.3.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#factors"><i class="fa fa-check"></i><b>7.3.4</b> Factors</a></li>
<li class="chapter" data-level="7.3.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-classtype"><i class="fa fa-check"></i><b>7.3.5</b> Checking class/type</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#basic-data-structures"><i class="fa fa-check"></i><b>7.4</b> Basic data structures</a></li>
<li class="chapter" data-level="7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>7.5</b> More information</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="expression-qc-and-normalization.html"><a href="expression-qc-and-normalization.html"><i class="fa fa-check"></i><b>8</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="9" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html"><i class="fa fa-check"></i><b>9</b> Data Wrangling scRNAseq</a></li>
<li class="chapter" data-level="10" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>10</b> Identifying Cell Populations</a></li>
<li class="chapter" data-level="11" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>11</b> Feature Selection and Cluster Analysis</a></li>
<li class="chapter" data-level="12" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>12</b> Batch Effects</a></li>
<li class="chapter" data-level="13" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html"><i class="fa fa-check"></i><b>13</b> Correcting Batch Effects</a></li>
<li class="chapter" data-level="14" data-path="functional-analysis.html"><a href="functional-analysis.html"><i class="fa fa-check"></i><b>14</b> Functional Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="functional-analysis.html"><a href="functional-analysis.html#gene-sets-and-signatures"><i class="fa fa-check"></i><b>14.1</b> Gene sets and signatures</a><ul>
<li class="chapter" data-level="14.1.1" data-path="functional-analysis.html"><a href="functional-analysis.html#cell-cycle"><i class="fa fa-check"></i><b>14.1.1</b> Cell Cycle</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="functional-analysis.html"><a href="functional-analysis.html#pathway-analysis"><i class="fa fa-check"></i><b>14.2</b> Pathway analysis</a></li>
<li class="chapter" data-level="14.3" data-path="functional-analysis.html"><a href="functional-analysis.html#infercnv-honeybadger"><i class="fa fa-check"></i><b>14.3</b> inferCNV / honeybadger</a><ul>
<li class="chapter" data-level="14.3.1" data-path="functional-analysis.html"><a href="functional-analysis.html#create-the-infercnv-object"><i class="fa fa-check"></i><b>14.3.1</b> Create the InferCNV Object</a></li>
<li class="chapter" data-level="14.3.2" data-path="functional-analysis.html"><a href="functional-analysis.html#filtering-genes"><i class="fa fa-check"></i><b>14.3.2</b> Filtering genes</a></li>
<li class="chapter" data-level="14.3.3" data-path="functional-analysis.html"><a href="functional-analysis.html#normalize-each-cells-counts-for-sequencing-depth"><i class="fa fa-check"></i><b>14.3.3</b> Normalize each cell’s counts for sequencing depth</a></li>
<li class="chapter" data-level="14.3.4" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-anscombe-normalization"><i class="fa fa-check"></i><b>14.3.4</b> Perform Anscombe normalization</a></li>
<li class="chapter" data-level="14.3.5" data-path="functional-analysis.html"><a href="functional-analysis.html#log-transform-the-normalized-counts"><i class="fa fa-check"></i><b>14.3.5</b> Log transform the normalized counts:</a></li>
<li class="chapter" data-level="14.3.6" data-path="functional-analysis.html"><a href="functional-analysis.html#apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects"><i class="fa fa-check"></i><b>14.3.6</b> Apply maximum bounds to the expression data to reduce outlier effects</a></li>
<li class="chapter" data-level="14.3.7" data-path="functional-analysis.html"><a href="functional-analysis.html#initial-view-before-infercnv-operations"><i class="fa fa-check"></i><b>14.3.7</b> Initial view, before inferCNV operations:</a></li>
<li class="chapter" data-level="14.3.8" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-smoothing-across-chromosomes"><i class="fa fa-check"></i><b>14.3.8</b> Perform smoothing across chromosomes</a></li>
<li class="chapter" data-level="14.3.9" data-path="functional-analysis.html"><a href="functional-analysis.html#subtract-the-reference-values-from-observations-now-have-logfold-change-values"><i class="fa fa-check"></i><b>14.3.9</b> Subtract the reference values from observations, now have log(fold change) values</a></li>
<li class="chapter" data-level="14.3.10" data-path="functional-analysis.html"><a href="functional-analysis.html#invert-log-values"><i class="fa fa-check"></i><b>14.3.10</b> Invert log values</a></li>
<li class="chapter" data-level="14.3.11" data-path="functional-analysis.html"><a href="functional-analysis.html#removing-noise"><i class="fa fa-check"></i><b>14.3.11</b> Removing noise</a></li>
<li class="chapter" data-level="14.3.12" data-path="functional-analysis.html"><a href="functional-analysis.html#remove-outlier-data-points"><i class="fa fa-check"></i><b>14.3.12</b> Remove outlier data points</a></li>
<li class="chapter" data-level="14.3.13" data-path="functional-analysis.html"><a href="functional-analysis.html#find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic"><i class="fa fa-check"></i><b>14.3.13</b> Find DE genes by comparing the mutant types to normal types, BASIC</a></li>
<li class="chapter" data-level="14.3.14" data-path="functional-analysis.html"><a href="functional-analysis.html#additional-information"><i class="fa fa-check"></i><b>14.3.14</b> Additional Information</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html"><i class="fa fa-check"></i><b>15</b> Pseudotime Cell Trajectories</a></li>
<li class="chapter" data-level="16" data-path="spatial-genomics.html"><a href="spatial-genomics.html"><i class="fa fa-check"></i><b>16</b> Spatial Genomics</a></li>
<li class="chapter" data-level="17" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html"><i class="fa fa-check"></i><b>17</b> Functional Pseudotime Analysis</a></li>
<li class="chapter" data-level="18" data-path="single-cell-multiomic-technologies.html"><a href="single-cell-multiomic-technologies.html"><i class="fa fa-check"></i><b>18</b> Single Cell Multiomic Technologies</a></li>
<li class="chapter" data-level="19" data-path="cite-seq-and-scatac-seq.html"><a href="cite-seq-and-scatac-seq.html"><i class="fa fa-check"></i><b>19</b> CITE-seq and scATAC-seq</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="chapter" data-level="20" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>20</b> Introduction</a></li>
<li class="chapter" data-level="21" data-path="literature.html"><a href="literature.html"><i class="fa fa-check"></i><b>21</b> Literature</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</h1>
<p class="author"><em>Orr Ashenberg</em></p>
<p class="author"><em>Dana Silverbush</em></p>
<p class="author"><em>Kirk Gosik</em></p>
<p class="date"><em>2019-02-21</em></p>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<div id="course-overview" class="section level2">
<h2><span class="header-section-number">1.1</span> COURSE OVERVIEW</h2>
<p>In recent years single cell RNA-seq (scRNA-seq) has become widely used for transcriptome analysis in many areas of biology. In contrast to bulk RNA-seq, scRNA-seq provides quantitative measurements of the expression of every gene in a single cell. However, to analyze scRNA-seq data, novel methods are required and some of the underlying assumptions for the methods developed for bulk RNA-seq experiments are no longer valid. In this course we will cover all steps of the scRNA-seq processing, starting from the raw reads coming off the sequencer. The course includes common analysis strategies, using state-of-the-art methods and we also discuss the central biological questions that can be addressed using scRNA-seq.</p>
</div>
<div id="targeted-audience-assumed-background" class="section level2">
<h2><span class="header-section-number">1.2</span> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</h2>
<p>This course is aimed at researchers and technical workers who are or will be analyzing scRNA-seq data. The material is suitable both for experimentalists who want to learn more about data-analysis as well as computational biologists who want to learn about scRNASeq methods. Examples demonstrated in this course can be applied to any experimental protocol or biological system.</p>
<p>The requirements for this course are: 1. Working knowledge of unix (managing files, running programs) 2. Programming experience in R (writing a function, basic I/O operations, variable types, using packages). Bioconductor experience is a plus. 3. Familiarity with next-generation sequencing data and its analyses (using alignment and quantification tools for bulk sequencing data)</p>
</div>
<div id="course-format" class="section level2">
<h2><span class="header-section-number">1.3</span> COURSE FORMAT</h2>
<p>The course will be delivered over the course of five days. Each day will include a lecture and laboratory component. The lecture will introduce the topics of discussion and the laboratory sessions will be focused on practical hands-on analysis of scRNA-seq data. These sessions will involve a combination of both mirroring exercises with the instructor to demonstrate a skill as well as applying these skills on your own to complete individual exercises. After and during each exercise, interpretation of results will be discussed as a group. Computing will be done using a combination of tools installed on the attendees laptop computer and web resources accessed via web browser.</p>
</div>
<div id="getting-started" class="section level2">
<h2><span class="header-section-number">1.4</span> Getting Started</h2>
</div>
<div id="session-content" class="section level2">
<h2><span class="header-section-number">1.5</span> SESSION CONTENT</h2>
<div id="monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks" class="section level3">
<h3><span class="header-section-number">1.5.1</span> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</h3>
<div id="lecture-1-scrna-seq-experimental-design-1-hour-orr" class="section level4">
<h4><span class="header-section-number">1.5.1.1</span> Lecture 1 – scRNA-Seq experimental design (1 hour, Orr)</h4>
<p>Overview of course (intros, 20 min) General introduction: HCA/KCO overview Comparison of Bulk and single cell RNA-Seq Overview of available scRNA-seq technologies (10x) and experimental protocols scRNA-Seq experimental design and analysis workflow?</p>
</div>
<div id="lab-1-understanding-sequencing-raw-data-kirk-1.5-2-hours" class="section level4">
<h4><span class="header-section-number">1.5.1.2</span> Lab 1 – Understanding sequencing raw data (Kirk, 1.5-2 hours)</h4>
<p>Lab based around data wrangling from public data repositories: get data from 10x website, single cell portal, from GEO (fastqs, counts) Shell and Unix commands to navigate directories, create folders, open files Raw file formats</p>
</div>
<div id="lecture-2---intro-to-data-processing-from-bcl-file-to-bam-file-1-hour-dana" class="section level4">
<h4><span class="header-section-number">1.5.1.3</span> Lecture 2 - Intro to Data processing: from bcl file to bam file (1 hour, Dana)</h4>
<p>scRNA-Seq processing workflow starting with choice of sequencer (NextSeq, HiSeq, MiSeq) / barcode swapping and bcl files Overview of Popular tools and algorithms Common single-cell analyses and interpretation Sequencing data: alignment and quality control Looking at cool things in alignment like where reads are, mutations, splicing</p>
</div>
<div id="lab-2-processing-raw-scrna-seq-data-dana-no-r-at-this-point" class="section level4">
<h4><span class="header-section-number">1.5.1.4</span> Lab 2 – Processing raw scRNA-Seq data (Dana, no R at this point)</h4>
<p>Data outputs from different scRNAseq technologies (10x, Smart-seq2) - process both? Demultiplexing sequencing data Read Quality Control (CellRanger, dropEst, fastqc) Run bowtie2 on 2 wells to demonstrate alignment Read alignment and visualization (kallisto, RSEM, Igviewer)</p>
<p><a href="http://hemberg-lab.github.io/scRNA.seq.course/processing-raw-scrna-seq-data.html" class="uri">http://hemberg-lab.github.io/scRNA.seq.course/processing-raw-scrna-seq-data.html</a> Chapters 3 and 4? Are there parts that supposed to be in labs 1 or 3?</p>
<ul>
<li>Demultiplexing</li>
<li>FastQC</li>
<li>Align (STAR/TOPHAT/Kallisto)</li>
<li>IGViewer - what do we want here? I use it for mutation detections, copying sequences, searching for alternative splicing.</li>
</ul>
<p>Flash talks (1.5 hr, break into 2 groups of 13) small presentation about your genome assembly and annotation project, ideally do 3 slides -2/3 mins (powerpoint or similar). So you can introduce yourselves and we can get to know each other.</p>
</div>
</div>
<div id="tuesday-classes-from-0930-to-1730" class="section level3">
<h3><span class="header-section-number">1.5.2</span> Tuesday – Classes from 09:30 to 17:30</h3>
<div id="lecture-3-transcriptome-quantification-from-bam-file-to-counts-orr" class="section level4">
<h4><span class="header-section-number">1.5.2.1</span> Lecture 3 – Transcriptome quantification: from bam file to counts (Orr)</h4>
<p>Read &amp; UMI counting (Kallisto alignment-free pseudocounts as well), how RSEM works (length dependence, sequencing depth, multimapping reads), CellRanger (dropest), bustools 10x barcode structure and links to Perturb-seq Gene length &amp; coverage Gene expression units (count data Smart-seq2 counts or 10x UMIs vs expression data) Some R overview slides, <a href="https://r4ds.had.co.nz/" class="uri">https://r4ds.had.co.nz/</a></p>
</div>
<div id="lab-3---introduction-to-r-kirk" class="section level4">
<h4><span class="header-section-number">1.5.2.2</span> Lab 3 - Introduction to R (Kirk)</h4>
<ul>
<li>Installing packages</li>
<li>Data-types</li>
<li>Data manipulation, slicing</li>
<li>Strings manipulations</li>
<li>Introducing object oriented programming / S4 objects</li>
<li>Visualization tools</li>
<li>Bonus create FeaturePlot from Seurat in base ggplot</li>
<li>Bonus: run RSEM on Dana’s bam files if you are bored</li>
</ul>
</div>
<div id="lecture-4---expression-qc-normalisation-and-batch-correction-dana" class="section level4">
<h4><span class="header-section-number">1.5.2.3</span> Lecture 4 - Expression QC, normalisation and batch correction (Dana)</h4>
<p>What CellRanger does for quality filtering PBMC data normalisation methods <a href="https://www.nature.com/articles/nmeth.4292" class="uri">https://www.nature.com/articles/nmeth.4292</a> Doublets, empty droplets Barcode swapping Regression with technical covariates What about imputation?</p>
</div>
<div id="lab-4-data-wrangling-for-scrnaseq-data-dana" class="section level4">
<h4><span class="header-section-number">1.5.2.4</span> Lab 4 – Data wrangling for scRNAseq data (Dana)</h4>
<p>Data structures and file formats for single-cell data Quality control of cells and genes (doublets, ambient, empty drops) Data exploration: violin plots…</p>
<p>Introducing Seurat object Genes House keeping genes Mitochondrial genes (never used these ones) Filter - Do we remove both cells and genes here? Normalize (introduce more options, other than log transform?) Find variable genes (Is it a first reduction? Why the binning?) Scaling Regression Heatmap of desired genes? Sigantures? <a href="http://hemberg-lab.github.io/scRNA.seq.course/introduction-to-rbioconductor.html" class="uri">http://hemberg-lab.github.io/scRNA.seq.course/introduction-to-rbioconductor.html</a> 9.1 - 9.5? Bonus - imputation (magic? One of the two Gocken recommended?)</p>
<p>Flash talks (1.5 hr, break into 2 groups of 13) small presentation about your genome assembly and annotation project, ideally do 3 slides -2/3 mins (powerpoint or similar). So you can introduce yourselves and we can get to know each other.</p>
</div>
</div>
<div id="wednesday-classes-from-0930-to-1730" class="section level3">
<h3><span class="header-section-number">1.5.3</span> Wednesday – Classes from 09:30 to 17:30</h3>
<div id="lecture-5---identifying-cell-populations-kirk" class="section level4">
<h4><span class="header-section-number">1.5.3.1</span> Lecture 5 - Identifying cell populations (Kirk)</h4>
<p>Feature selection Dimensionality reduction Clustering and assigning identity (Louvain, NMF, topic models, variational autoencoder) Differential expression tests</p>
</div>
<div id="lab-5-feature-selection-clustering-analysis-kirk" class="section level4">
<h4><span class="header-section-number">1.5.3.2</span> Lab 5 – Feature selection &amp; Clustering analysis (Kirk)</h4>
<p>Parameters and clustering Comparison of feature selection methods</p>
</div>
<div id="lecture-6---introduction-to-batch-effects-orr" class="section level4">
<h4><span class="header-section-number">1.5.3.3</span> Lecture 6 - Introduction to batch effects (Orr)</h4>
<p>Batch correction methods (regress out batch, scaling within batch, CCA, MNN, Liger, scvi, scgen) Evaluation methods for batch correction (ARI, average silhouette width, kBET…)</p>
</div>
<div id="lab-6---correcting-batch-effects-orr" class="section level4">
<h4><span class="header-section-number">1.5.3.4</span> Lab 6 - Correcting batch effects (Orr)</h4>
<p>Comparison of batch correction methods, Seurat pancreas</p>
<p>Poster session with beer &amp; wine (time?) poster of current or proposed single cell genomic study. Print out would be good (recent posters are fine), or if you haven’t got a recent one, just a quick one pulled together on powerpoint and printed out on A4 is okay too - you shouldn’t stress too much about this, it’s just to connect the flash talks to the posters and it will be absolutely informal (unlike a poster session at a conference!). Take notes on how to do single-cell analysis, ideas, challenges, things you find interesting, directions you would like to explore.</p>
</div>
</div>
<div id="thursday-classes-from-0930-to-1730" class="section level3">
<h3><span class="header-section-number">1.5.4</span> Thursday – Classes from 09:30 to 17:30</h3>
<p>Post poster session discussion groups 30 min discussion wrapping up poster session, and keeping track of ideas in a shared Google doc</p>
<div id="lecture-7---advanced-topics-functional-analysis-of-cell-sub-populations-kirk-30-min" class="section level4">
<h4><span class="header-section-number">1.5.4.1</span> Lecture 7 - Advanced topics: Functional analysis of cell sub-populations (Kirk 30 min)</h4>
<p>Gene sets and signatures Pathway analysis inferCNV / honeybadger</p>
</div>
<div id="lecture-7---advanced-topics-pseudotime-cell-trajectories-kirk-30-min" class="section level4">
<h4><span class="header-section-number">1.5.4.2</span> Lecture 7 - Advanced topics: Pseudotime cell trajectories (Kirk 30 min)</h4>
<ul>
<li>Waddington Landscape</li>
<li>Pseudotime inference</li>
<li><p>Differential expression through pseudotime</p></li>
<li><p>Lecture 7 - Advanced topics: Spatial genomics, multiplexing? (Kirk 30 min)</p></li>
</ul>
</div>
<div id="lab-8---functional-and-pseudotime-analysis-orr" class="section level4">
<h4><span class="header-section-number">1.5.4.3</span> Lab 8 - Functional and Pseudotime analysis (Orr)</h4>
<p>Popular tools and packages for functional analysis (<a href="https://github.com/dynverse/dynmethods#list-of-included-methods" class="uri">https://github.com/dynverse/dynmethods#list-of-included-methods</a>) Review concepts from papers Comparison of pseudotime methods <a href="https://github.com/velocyto-team/velocyto.R" class="uri">https://github.com/velocyto-team/velocyto.R</a></p>
</div>
<div id="lecture-8---single-cell-multiomic-technologies-dana" class="section level4">
<h4><span class="header-section-number">1.5.4.4</span> Lecture 8 - Single-cell multiomic technologies (Dana)</h4>
<p>Introduction to other omic data types Integrating scRNA-seq with other single-cell modalities (CITE, Perturb, ATAC, methylation…)</p>
</div>
<div id="lab-9---analysis-of-cite-seq-scatac-seq-orr-may-be-time-limited" class="section level4">
<h4><span class="header-section-number">1.5.4.5</span> Lab 9 - Analysis of CITE-seq, scATAC-seq (Orr, may be time-limited)</h4>
<ul>
<li><a href="https://github.com/Hoohm/CITE-seq-Count" class="uri">https://github.com/Hoohm/CITE-seq-Count</a></li>
<li><a href="https://cite-seq.com/eccite-seq/" class="uri">https://cite-seq.com/eccite-seq/</a></li>
<li><a href="https://support.10xgenomics.com/single-cell-vdj/index/doc/technical-note-assay-scheme-and-configuration-of-chromium-single-cell-vdj-libraries" class="uri">https://support.10xgenomics.com/single-cell-vdj/index/doc/technical-note-assay-scheme-and-configuration-of-chromium-single-cell-vdj-libraries</a></li>
<li><a href="https://satijalab.org/seurat/multimodal_vignette.html" class="uri">https://satijalab.org/seurat/multimodal_vignette.html</a></li>
<li><a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/cicero/inst/doc/website.html" class="uri">https://www.bioconductor.org/packages/devel/bioc/vignettes/cicero/inst/doc/website.html</a></li>
</ul>
</div>
</div>
<div id="friday-classes-from-0930-to-1730" class="section level3">
<h3><span class="header-section-number">1.5.5</span> Friday – Classes from 09:30 to 17:30</h3>
<div id="lab-10---small-dataset-for-analysis-dana" class="section level4">
<h4><span class="header-section-number">1.5.5.1</span> Lab 10 - small dataset for analysis (Dana)</h4>
<p>Karthik has a good starting point: Krumlov_lab2_guidelines.html and Hemberg lab <a href="http://hemberg-lab.github.io/scRNA.seq.course/advanced-exercises.html" class="uri">http://hemberg-lab.github.io/scRNA.seq.course/advanced-exercises.html</a> Present a set of methods and orders in which to try them IDH mutated glioma Goal for first half: get to clustering and identify malignant cells (inferCNV possible) Goal for second half: cell states For our other courses, the last day we usually divide them in groups of 3-4 ppl and assign them a small dataset in order to repeat all the analyses they have learnt during the week, to think about the best strategy to carry out the analyses and to present the results in front of the other ppl. I believe this is something very productive and helpful because it is a kind of wrap-up and brainstorming session. Normally, ppl love to talk and present something and in this way they/we can be 100% sure they really understood the different steps of the pipeline.</p>
<p>Lecture 9 - Group presentations Review, Questions and Answers</p>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="scrna-seq-experimental-design.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["2019_scWorkshop.pdf", "2019_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
