<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>7 Introduction R/Bioconductor | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="7 Introduction R/Bioconductor | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2019_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="7 Introduction R/Bioconductor | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg">
<meta name="author" content="Dana Silverbush">
<meta name="author" content="Kirk Gosik">


<meta name="date" content="2019-05-30">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="transcriptome-quantification.html">
<link rel="next" href="expression-qc-and-normalization.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks"><i class="fa fa-check"></i><b>1.5.1</b> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 09:30 to 17:30</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scrna-seq-experimental-design.html"><a href="scrna-seq-experimental-design.html"><i class="fa fa-check"></i><b>2</b> scRNA-Seq Experimental Design</a></li>
<li class="chapter" data-level="3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>3</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#class-environment"><i class="fa fa-check"></i><b>3.1</b> Class Environment</a><ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#getting-into-aws-instance"><i class="fa fa-check"></i><b>3.1.1</b> Getting into AWS Instance</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#shell-and-unix-commands"><i class="fa fa-check"></i><b>3.2</b> Shell and Unix commands</a><ul>
<li class="chapter" data-level="3.2.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>3.2.1</b> Common Linux Commands</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#file-formats"><i class="fa fa-check"></i><b>3.3</b> File formats</a><ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-fastq-files"><i class="fa fa-check"></i><b>3.3.1</b> View FASTQ Files</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-bam-files"><i class="fa fa-check"></i><b>3.3.2</b> View BAM Files</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#public-data-repositories"><i class="fa fa-check"></i><b>3.4</b> Public data repositories</a><ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#cellranger10x"><i class="fa fa-check"></i><b>3.4.1</b> Cellranger/10x</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>3.4.2</b> GEO</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>3.4.3</b> Single Cell Portal</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Data Preprocessing</a></li>
<li class="chapter" data-level="5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html"><i class="fa fa-check"></i><b>5</b> Processing scRNAseq Data</a><ul>
<li class="chapter" data-level="5.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#goal"><i class="fa fa-check"></i><b>5.1</b> Goal</a></li>
<li class="chapter" data-level="5.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#further-reading"><i class="fa fa-check"></i><b>5.2</b> Further reading</a></li>
<li class="chapter" data-level="5.3" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastqc"><i class="fa fa-check"></i><b>5.3</b> FastQC</a><ul>
<li class="chapter" data-level="5.3.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastq-file-format"><i class="fa fa-check"></i><b>5.3.1</b> Fastq file format</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#align-the-reads"><i class="fa fa-check"></i><b>5.4</b> Align the reads</a><ul>
<li class="chapter" data-level="5.4.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#star-align"><i class="fa fa-check"></i><b>5.4.1</b> STAR align</a></li>
<li class="chapter" data-level="5.4.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#bam-file-format"><i class="fa fa-check"></i><b>5.4.2</b> Bam file format</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#visualization"><i class="fa fa-check"></i><b>5.5</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="transcriptome-quantification.html"><a href="transcriptome-quantification.html"><i class="fa fa-check"></i><b>6</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>7</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#start-environment"><i class="fa fa-check"></i><b>7.1</b> Start Environment</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>7.2</b> Installing packages</a><ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>7.2.1</b> CRAN</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>7.2.2</b> Github</a></li>
<li class="chapter" data-level="7.2.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>7.2.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.2.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>7.2.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>7.3</b> Installation instructions:</a><ul>
<li class="chapter" data-level="7.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#classestypes"><i class="fa fa-check"></i><b>7.3.1</b> Classes/Types</a></li>
<li class="chapter" data-level="7.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-structures"><i class="fa fa-check"></i><b>7.3.2</b> Data structures</a></li>
<li class="chapter" data-level="7.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#detour-to-s3s4"><i class="fa fa-check"></i><b>7.3.3</b> Detour to S3/S4</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>7.4</b> More information</a><ul>
<li class="chapter" data-level="7.4.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-for-help-for-any-function"><i class="fa fa-check"></i><b>7.4.1</b> Checking for help for any function!</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#grammer-of-graphics-ggplot2"><i class="fa fa-check"></i><b>7.5</b> Grammer of Graphics (ggplot2)</a><ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>7.5.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>7.5.2</b> Principles of ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#reference"><i class="fa fa-check"></i><b>7.6</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="expression-qc-and-normalization.html"><a href="expression-qc-and-normalization.html"><i class="fa fa-check"></i><b>8</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="9" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html"><i class="fa fa-check"></i><b>9</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="9.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#goal-1"><i class="fa fa-check"></i><b>9.1</b> Goal</a></li>
<li class="chapter" data-level="9.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#introduction-1"><i class="fa fa-check"></i><b>9.2</b> Introduction</a><ul>
<li class="chapter" data-level="9.2.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#load-necessary-packages"><i class="fa fa-check"></i><b>9.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="9.2.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>9.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="9.2.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>9.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="9.2.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>9.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="9.2.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>9.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>9.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="9.3.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>9.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="9.3.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>9.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#beginning-with-seurat-httpsatijalab.orgseurat"><i class="fa fa-check"></i><b>9.4</b> Beginning with Seurat: <a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a></a><ul>
<li class="chapter" data-level="9.4.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>9.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>9.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="9.6" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>9.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="9.6.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>9.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>9.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="9.8" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>9.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>10</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="10.1" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html#google-slides"><i class="fa fa-check"></i><b>10.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>11</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#abstract"><i class="fa fa-check"></i><b>11.1</b> Abstract</a></li>
<li class="chapter" data-level="11.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#seurat-tutorial"><i class="fa fa-check"></i><b>11.2</b> Seurat Tutorial</a><ul>
<li class="chapter" data-level="11.2.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#preprocessing-steps"><i class="fa fa-check"></i><b>11.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="11.2.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>11.2.2</b> Start of Identifying Cell Types</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#feature-selection"><i class="fa fa-check"></i><b>11.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="11.3.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>11.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="11.3.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>11.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="11.3.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#clustering"><i class="fa fa-check"></i><b>11.3.4</b> Clustering</a></li>
<li class="chapter" data-level="11.3.5" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#check-clusters"><i class="fa fa-check"></i><b>11.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="11.3.6" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>11.3.6</b> Practice Visualizing/Embedding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>12</b> Batch Effects</a></li>
<li class="chapter" data-level="13" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html"><i class="fa fa-check"></i><b>13</b> Correcting Batch Effects</a><ul>
<li class="chapter" data-level="13.1" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html#load-settings-and-packages"><i class="fa fa-check"></i><b>13.1</b> Load settings and packages</a></li>
<li class="chapter" data-level="13.2" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html#preparing-the-individual-seurat-objects-for-each-pancreas-dataset-without-batch-correction"><i class="fa fa-check"></i><b>13.2</b> Preparing the individual Seurat objects for each pancreas dataset without batch correction</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html"><i class="fa fa-check"></i><b>14</b> Cluster pancreatic datasets without batch correction</a><ul>
<li class="chapter" data-level="14.0.1" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#batch-correction-canonical-correlation-analysis-cca-using-seurat"><i class="fa fa-check"></i><b>14.0.1</b> Batch correction: canonical correlation analysis (CCA) using Seurat</a></li>
<li class="chapter" data-level="14.0.2" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#batch-correction-integrative-non-negative-matrix-factorization-nmf-using-liger"><i class="fa fa-check"></i><b>14.0.2</b> Batch correction: integrative non-negative matrix factorization (NMF) using LIGER</a></li>
<li class="chapter" data-level="14.1" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#additional-exploration-regressing-out-unwanted-covariates"><i class="fa fa-check"></i><b>14.1</b> Additional exploration: Regressing out unwanted covariates</a></li>
<li class="chapter" data-level="14.2" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#additional-exploration-kbet"><i class="fa fa-check"></i><b>14.2</b> Additional exploration: kBET</a></li>
<li class="chapter" data-level="14.3" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#additional-exploration-seurat-3"><i class="fa fa-check"></i><b>14.3</b> Additional exploration: Seurat 3</a></li>
<li class="chapter" data-level="14.4" data-path="cluster-pancreatic-datasets-without-batch-correction.html"><a href="cluster-pancreatic-datasets-without-batch-correction.html#acknowledgements"><i class="fa fa-check"></i><b>14.4</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="functional-analysis.html"><a href="functional-analysis.html"><i class="fa fa-check"></i><b>15</b> Functional Analysis</a><ul>
<li class="chapter" data-level="15.1" data-path="functional-analysis.html"><a href="functional-analysis.html#google-slides-1"><i class="fa fa-check"></i><b>15.1</b> Google Slides</a></li>
<li class="chapter" data-level="15.2" data-path="functional-analysis.html"><a href="functional-analysis.html#gene-sets-and-signatures"><i class="fa fa-check"></i><b>15.2</b> Gene sets and signatures</a><ul>
<li class="chapter" data-level="15.2.1" data-path="functional-analysis.html"><a href="functional-analysis.html#cell-cycle"><i class="fa fa-check"></i><b>15.2.1</b> Cell Cycle</a></li>
</ul></li>
<li class="chapter" data-level="15.3" data-path="functional-analysis.html"><a href="functional-analysis.html#pathway-analysis"><i class="fa fa-check"></i><b>15.3</b> Pathway analysis</a></li>
<li class="chapter" data-level="15.4" data-path="functional-analysis.html"><a href="functional-analysis.html#infercnv-honeybadger"><i class="fa fa-check"></i><b>15.4</b> inferCNV / honeybadger</a><ul>
<li class="chapter" data-level="15.4.1" data-path="functional-analysis.html"><a href="functional-analysis.html#create-the-infercnv-object"><i class="fa fa-check"></i><b>15.4.1</b> Create the InferCNV Object</a></li>
<li class="chapter" data-level="15.4.2" data-path="functional-analysis.html"><a href="functional-analysis.html#filtering-genes"><i class="fa fa-check"></i><b>15.4.2</b> Filtering genes</a></li>
<li class="chapter" data-level="15.4.3" data-path="functional-analysis.html"><a href="functional-analysis.html#normalize-each-cells-counts-for-sequencing-depth"><i class="fa fa-check"></i><b>15.4.3</b> Normalize each cell’s counts for sequencing depth</a></li>
<li class="chapter" data-level="15.4.4" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-anscombe-normalization"><i class="fa fa-check"></i><b>15.4.4</b> Perform Anscombe normalization</a></li>
<li class="chapter" data-level="15.4.5" data-path="functional-analysis.html"><a href="functional-analysis.html#log-transform-the-normalized-counts"><i class="fa fa-check"></i><b>15.4.5</b> Log transform the normalized counts:</a></li>
<li class="chapter" data-level="15.4.6" data-path="functional-analysis.html"><a href="functional-analysis.html#apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects"><i class="fa fa-check"></i><b>15.4.6</b> Apply maximum bounds to the expression data to reduce outlier effects</a></li>
<li class="chapter" data-level="15.4.7" data-path="functional-analysis.html"><a href="functional-analysis.html#initial-view-before-infercnv-operations"><i class="fa fa-check"></i><b>15.4.7</b> Initial view, before inferCNV operations:</a></li>
<li class="chapter" data-level="15.4.8" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-smoothing-across-chromosomes"><i class="fa fa-check"></i><b>15.4.8</b> Perform smoothing across chromosomes</a></li>
<li class="chapter" data-level="15.4.9" data-path="functional-analysis.html"><a href="functional-analysis.html#subtract-the-reference-values-from-observations-now-have-logfold-change-values"><i class="fa fa-check"></i><b>15.4.9</b> Subtract the reference values from observations, now have log(fold change) values</a></li>
<li class="chapter" data-level="15.4.10" data-path="functional-analysis.html"><a href="functional-analysis.html#invert-log-values"><i class="fa fa-check"></i><b>15.4.10</b> Invert log values</a></li>
<li class="chapter" data-level="15.4.11" data-path="functional-analysis.html"><a href="functional-analysis.html#removing-noise"><i class="fa fa-check"></i><b>15.4.11</b> Removing noise</a></li>
<li class="chapter" data-level="15.4.12" data-path="functional-analysis.html"><a href="functional-analysis.html#remove-outlier-data-points"><i class="fa fa-check"></i><b>15.4.12</b> Remove outlier data points</a></li>
<li class="chapter" data-level="15.4.13" data-path="functional-analysis.html"><a href="functional-analysis.html#find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic"><i class="fa fa-check"></i><b>15.4.13</b> Find DE genes by comparing the mutant types to normal types, BASIC</a></li>
<li class="chapter" data-level="15.4.14" data-path="functional-analysis.html"><a href="functional-analysis.html#additional-information"><i class="fa fa-check"></i><b>15.4.14</b> Additional Information</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="16" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html"><i class="fa fa-check"></i><b>16</b> Pseudotime Cell Trajectories</a><ul>
<li class="chapter" data-level="16.1" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#google-slides-2"><i class="fa fa-check"></i><b>16.1</b> Google Slides</a></li>
<li class="chapter" data-level="16.2" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#comparison-abstract"><i class="fa fa-check"></i><b>16.2</b> Comparison Abstract</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html"><i class="fa fa-check"></i><b>17</b> Functional Pseudotime Analysis</a><ul>
<li class="chapter" data-level="17.1" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#load-settings-and-packages-1"><i class="fa fa-check"></i><b>17.1</b> Load settings and packages</a></li>
<li class="chapter" data-level="17.2" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#first-look-at-the-differentiation-data-from-deng-et-al."><i class="fa fa-check"></i><b>17.2</b> First look at the differentiation data from Deng et al.</a></li>
<li class="chapter" data-level="17.3" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#diffusion-map-pseudotime"><i class="fa fa-check"></i><b>17.3</b> Diffusion map pseudotime</a></li>
<li class="chapter" data-level="17.4" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#slingshot-map-pseudotime"><i class="fa fa-check"></i><b>17.4</b> Slingshot map pseudotime</a></li>
<li class="chapter" data-level="17.5" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#find-temporally-expressed-genes"><i class="fa fa-check"></i><b>17.5</b> Find temporally expressed genes</a></li>
<li class="chapter" data-level="17.6" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#comparison-of-the-different-trajectory-inference-methods"><i class="fa fa-check"></i><b>17.6</b> Comparison of the different trajectory inference methods</a></li>
<li class="chapter" data-level="17.7" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#plots-of-gene-expression-over-time."><i class="fa fa-check"></i><b>17.7</b> Plots of gene expression over time.</a></li>
<li class="chapter" data-level="17.8" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html#acknowledgements-1"><i class="fa fa-check"></i><b>17.8</b> Acknowledgements</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="single-cell-multiomic-technologies.html"><a href="single-cell-multiomic-technologies.html"><i class="fa fa-check"></i><b>18</b> Single Cell Multiomic Technologies</a></li>
<li class="chapter" data-level="19" data-path="cite-seq-and-scatac-seq.html"><a href="cite-seq-and-scatac-seq.html"><i class="fa fa-check"></i><b>19</b> CITE-seq and scATAC-seq</a></li>
<li class="chapter" data-level="20" data-path="single-cell-resources.html"><a href="single-cell-resources.html"><i class="fa fa-check"></i><b>20</b> Single Cell Resources</a><ul>
<li class="chapter" data-level="20.1" data-path="single-cell-resources.html"><a href="single-cell-resources.html#comprehensive-list-of-single-cell-resources"><i class="fa fa-check"></i><b>20.1</b> Comprehensive list of single-cell resources</a></li>
<li class="chapter" data-level="20.2" data-path="single-cell-resources.html"><a href="single-cell-resources.html#computational-packages-for-single-cell-analysis"><i class="fa fa-check"></i><b>20.2</b> Computational packages for single-cell analysis</a></li>
<li class="chapter" data-level="20.3" data-path="single-cell-resources.html"><a href="single-cell-resources.html#elife-commentary-on-the-human-cell-atlas"><i class="fa fa-check"></i><b>20.3</b> eLife Commentary on the Human Cell Atlas</a></li>
<li class="chapter" data-level="20.4" data-path="single-cell-resources.html"><a href="single-cell-resources.html#online-courses"><i class="fa fa-check"></i><b>20.4</b> Online courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="introduction-rbioconductor" class="section level1">
<h1><span class="header-section-number">7</span> Introduction R/Bioconductor</h1>
<div id="start-environment" class="section level2">
<h2><span class="header-section-number">7.1</span> Start Environment</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## maybe take away the --rm so they can save the container for later</span>
<span class="co">## run from your home directory</span>
<span class="bu">cd</span> 

<span class="co">## example for user99</span>
<span class="ex">docker</span> run --rm -it -e PASSWORD=train \
  -v <span class="va">$PWD</span>/Share:/Share \
  -v <span class="va">$PWD</span>:/mydir \
  -p 9099:8787 kdgosik/scellbern2019</code></pre></div>
<p><strong>Explaination of commands</strong></p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">  <span class="ex">-</span> docker: command to run docker
  <span class="ex">-</span> run: asking docker to run a container
  <span class="ex">-</span> --rm: flag to remove the container when you exit from it
      <span class="ex">-</span> nothing will be saved from your session to access again later
      <span class="ex">-</span> this flag can be removed to keep container
  <span class="ex">-</span> -it: flag to run the container interactively
  <span class="ex">-</span> - this will keep all session output displaying on the terminal
  <span class="ex">-</span> - to stop container go to terminal and press Crtl+c
  <span class="ex">-</span> -v <span class="va">$PWD</span>/Share:/Share: map the share directory from AWS to Share inside docker container
  <span class="ex">-</span> -v <span class="va">$PWD</span>:/mydir: map your home directory to a directory inside docker container called home
  <span class="ex">-</span> -p 9017:8787: map docker container port of 8787(rstudio port default) <span class="ex">to</span> your computer port 9017
  <span class="ex">-</span> kdgosik/scellbern2019: the image to run.  It will be the image into a container if not already built on your computer
    <span class="ex">-</span> [image link](https://hub.docker.com/r/kdgosik/scellbern2019)</code></pre></div>
<ul>
<li><a href="https://localhost:9099">localhost:9099</a> or on AWS</li>
<li><p><a href="https://:9099"><AWS PUBLIC IP ADDRESS>:9099</a></p></li>
<li>ec2-<AWS PUBLIC IP ADDRESS>.us-west-2.compute.amazonaws.com:$PORT_NUMBER</li>
<li><p>ec2-54-202-32-102.us-west-2.compute.amazonaws.com:9099</p></li>
<li>R/Rstudio parts</li>
<li>Data Types and classes</li>
<li>Packages and where to get them</li>
<li>S3 vs S4</li>
<li><p>Visualizations and ggplot</p></li>
<li>Installing packages</li>
<li>Data-types</li>
<li>Data manipulation, slicing</li>
<li>Strings manipulations</li>
<li>Introducing object oriented programming / S4 objects</li>
<li>Visualization tools</li>
<li>Bonus create FeaturePlot from Seurat in base ggplot</li>
<li><p>Bonus: run RSEM on Dana’s bam files if you are bored</p></li>
</ul>
</div>
<div id="installing-packages" class="section level2">
<h2><span class="header-section-number">7.2</span> Installing packages</h2>
<div id="cran" class="section level3">
<h3><span class="header-section-number">7.2.1</span> CRAN</h3>
<p>The Comprehensive R Archive Network <a href="https://cran.r-project.org/">CRAN</a> is the biggest archive of R packages. There are few requirements for uploading packages besides building and installing succesfully, hence documentation and support is often minimal and figuring how to use these packages can be a challenge it itself. CRAN is the default repository R will search to find packages to install:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)

<span class="co"># or multiple packages</span>

<span class="kw">install.packages</span>(<span class="kw">c</span>(<span class="st">&quot;ggplot2&quot;</span>, <span class="st">&quot;stringr&quot;</span>))</code></pre></div>
</div>
<div id="github" class="section level3">
<h3><span class="header-section-number">7.2.2</span> Github</h3>
<p><a href="https://github.com/">Github</a> isn’t specific to R, any code of any type in any state can be uploaded. There is no guarantee a package uploaded to github will even install, nevermind do what it claims to do. R packages can be downloaded and installed directly from github using the “devtools” package installed above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## username/repository
devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;satijalab/seurat&quot;</span>) <span class="co"># latest stable version of Seurat package</span></code></pre></div>
<p>Github is also a version control system which stores multiple versions of any package. By default the most recent “master” version of the package is installed. If you want an older version or the development branch this can be specified using the “ref” parameter:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># different branch</span>
devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;satijalab/seurat&quot;</span>, <span class="dt">ref=</span><span class="st">&quot;release3.0&quot;</span>)

<span class="co"># previous commit</span>
## Merge branch &#39;develop&#39; into feat/MultiModal
##  - Shiwei Zheng committed on Jul 2, 2018
devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;tallulandrews/M3Drop&quot;</span>, <span class="dt">ref=</span><span class="st">&quot;551014f488770627ab154a62e59d49df5df98a3f&quot;</span>)</code></pre></div>
<p>Note: make sure you re-install the M3Drop master branch for later in the course.</p>
</div>
<div id="bioconductor" class="section level3">
<h3><span class="header-section-number">7.2.3</span> Bioconductor</h3>
<p>Bioconductor is a repository of R-packages specifically for biological analyses. It has the strictest requirements for submission, including installation on every platform and full documentation with a tutorial (called a vignette) explaining how the package should be used. Bioconductor also encourages utilization of standard data structures/classes and coding style/naming conventions, so that, in theory, packages and analyses can be combined into large pipelines or workflows.</p>
<p>Bioconductor also requires creators to support their packages and has a regular 6-month release schedule. Make sure you are using the most recent release of bioconductor before trying to install packages for the course.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## &gt;= R 3.5.0
<span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;Rsamtools&quot;</span>, <span class="dt">version =</span> <span class="st">&quot;3.8&quot;</span>, <span class="dt">ask =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
<div id="source" class="section level3">
<h3><span class="header-section-number">7.2.4</span> Source</h3>
<p>The final way to install packages is directly from source. In this case you have to download a fully built source code file, usually packagename.tar.gz, or clone the github repository and rebuild the package yourself. Generally this will only be done if you want to edit a package yourself, or if for some reason the former methods have failed. You can also get previous packages that aren’t supported any more on the <a href="https://cran.r-project.org/">CRAN package archive</a></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## Get an old package and install from source
<span class="kw">install.packages</span>(<span class="st">&quot;GenABEL_1.8-0.tar.gz&quot;</span>, <span class="dt">type=</span><span class="st">&quot;source&quot;</span>)</code></pre></div>
</div>
</div>
<div id="installation-instructions" class="section level2">
<h2><span class="header-section-number">7.3</span> Installation instructions:</h2>
<p>All the packages necessary for this course are available <a href="https://github.com/broadinstitute/2019_scWorkshop">here</a>. A list of the packages will be on the README.md for the repository. A script is also available inside the docker/install.R file.</p>
<div id="classestypes" class="section level3">
<h3><span class="header-section-number">7.3.1</span> Classes/Types</h3>
<p>R is a high level language so the underlying data-type is generally not important. The exception if you are accessing R data directly using another language such as C, but that is beyond the scope of this course. Instead we will consider the basic data classes: numeric, integer, logical, and character, and the higher level data class called “factor”. You can check what class your data is using the “class()” function.</p>
<div id="integer" class="section level4">
<h4><span class="header-section-number">7.3.1.1</span> Integer</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="dv">4</span> ## assign value of 4 to x

<span class="kw">class</span>(x) ## check class of x</code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.integer</span>(x) ## check if x is an integer</code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.numeric</span>(x) ## check if x is numeric</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(x) ## assign y to be an numeric

<span class="kw">is.numeric</span>(x) ## check if the assignment worked</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(x) ## check if the assignment worked</code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x ## check value of x</code></pre></div>
<pre><code>## [1] 4</code></pre>
</div>
<div id="numeric" class="section level4">
<h4><span class="header-section-number">7.3.1.2</span> Numeric</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## assign value of 1.414 to y
y &lt;-<span class="st"> </span><span class="fl">1.414</span>
## check class of y
<span class="kw">class</span>(y)</code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## check if y is numeric
<span class="kw">is.numeric</span>(y)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## check if y is an integer
<span class="kw">is.integer</span>(y)</code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## assign y to be an integer
y &lt;-<span class="st"> </span><span class="kw">as.integer</span>(y)
## check if the assignment worked
<span class="kw">is.integer</span>(y)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## check value of y
y</code></pre></div>
<pre><code>## [1] 1</code></pre>
</div>
<div id="logical-boolean" class="section level4">
<h4><span class="header-section-number">7.3.1.3</span> Logical/ Boolean</h4>
<p>The <code>logical</code> class stores boolean truth values, i.e. TRUE and FALSE. It is used for storing the results of logical operations and conditional statements will be coerced to this class. Most other data-types can be coerced to boolean without triggering (or “throwing”) error messages, which may cause unexpected behaviour.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">z &lt;-<span class="st"> </span><span class="ot">TRUE</span> ## assign value of TRUE to z
<span class="kw">class</span>(z) ## check class of z</code></pre></div>
<pre><code>## [1] &quot;logical&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.logical</span>(z) ## check if z is of logical type</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="data-structures" class="section level3">
<h3><span class="header-section-number">7.3.2</span> Data structures</h3>
<ul>
<li>Homogeneous
<ul>
<li>1D: atomic vector</li>
<li>2D: matrix</li>
<li>nD: array</li>
</ul></li>
<li>Heterogeneous
<ul>
<li>1D: list</li>
<li>2D: data.frame</li>
</ul></li>
</ul>
<div id="character-vectors" class="section level4">
<h4><span class="header-section-number">7.3.2.1</span> Character Vectors</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## assign a character vector with c() operator
character_vector &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;C&quot;</span>)

## check class
<span class="kw">class</span>(character_vector)</code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## access the 3rd element with [] operator
## *note*: R is index starts at 1 (other programming languages start at 0)
character_vector[<span class="dv">3</span>]</code></pre></div>
<pre><code>## [1] &quot;T&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## access 3rd through 6th elemenet
character_vector[<span class="dv">3</span><span class="op">:</span><span class="dv">6</span>]</code></pre></div>
<pre><code>## [1] &quot;T&quot; &quot;G&quot; &quot;C&quot; &quot;T&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## access the elements 1,4,7,10 with c()
character_vector[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">4</span>, <span class="dv">7</span>, <span class="dv">10</span>)]</code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;G&quot; &quot;G&quot; &quot;A&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## access all the A&#39;s
character_vector[<span class="kw">grep</span>(<span class="st">&quot;A&quot;</span>, character_vector)]</code></pre></div>
<pre><code>## [1] &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot;</code></pre>
</div>
<div id="numeric-vector" class="section level4">
<h4><span class="header-section-number">7.3.2.2</span> Numeric Vector</h4>
<p>The “numeric” class is the default class for storing any numeric data - integers, decimal numbers, numbers in scientific notation, etc…</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## assign a character vector with c() operator
numeric_vector &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">21</span>, <span class="dv">17</span>, <span class="dv">98</span>, <span class="dv">35</span>, <span class="dv">11</span>, <span class="dv">13</span>)

## check class
<span class="kw">class</span>(numeric_vector)</code></pre></div>
<pre><code>## [1] &quot;numeric&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## access the 5th element with [] operator
numeric_vector[<span class="dv">5</span>]</code></pre></div>
<pre><code>## [1] 98</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## access 2nd through 4th elemenet
numeric_vector[<span class="dv">2</span><span class="op">:</span><span class="dv">4</span>]</code></pre></div>
<pre><code>## [1]  5 21 17</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## backticks ` ` allow you to give names with non-typical characters
<span class="st">`</span><span class="dt">numeric?_vecotr</span><span class="st">`</span> &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;A&quot;</span>, <span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">21</span>, <span class="dv">17</span>, <span class="dv">98</span>, <span class="dv">35</span>, <span class="dv">11</span>, <span class="dv">13</span>)

## check vector
<span class="st">`</span><span class="dt">numeric?_vecotr</span><span class="st">`</span></code></pre></div>
<pre><code>## [1] &quot;A&quot;  &quot;1&quot;  &quot;5&quot;  &quot;21&quot; &quot;17&quot; &quot;98&quot; &quot;35&quot; &quot;11&quot; &quot;13&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## check class (Notice the quotation marks on the numbers!)
<span class="kw">class</span>(<span class="st">`</span><span class="dt">numeric?_vecotr</span><span class="st">`</span>)</code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
</div>
<div id="factor-vector" class="section level4">
<h4><span class="header-section-number">7.3.2.3</span> Factor Vector</h4>
<p>String/Character data is very memory inefficient to store, each letter generally requires the same amount of memory as any integer. Thus when storing a vector of strings with repeated elements it is more efficient assign each element to an integer and store the vector as integers and an additional string-to-integer association table. Thus, by default R will read in text columns of a data table as factors.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">factor_vector &lt;-<span class="st"> </span><span class="kw">factor</span>(numeric_vector)
factor_vector</code></pre></div>
<pre><code>## [1] 1  5  21 17 98 35 11 13
## Levels: 1 5 11 13 17 21 35 98</code></pre>
</div>
<div id="named-vector" class="section level4">
<h4><span class="header-section-number">7.3.2.4</span> Named Vector</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(numeric_vector) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Patient&quot;</span>, <span class="dv">1</span> <span class="op">:</span><span class="st"> </span><span class="kw">length</span>(numeric_vector))
numeric_vector</code></pre></div>
<pre><code>## Patient1 Patient2 Patient3 Patient4 Patient5 Patient6 Patient7 Patient8 
##        1        5       21       17       98       35       11       13</code></pre>
</div>
<div id="list" class="section level4">
<h4><span class="header-section-number">7.3.2.5</span> List</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## change the c() operator to list() operator
new_list &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;A&quot;</span>, <span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">21</span>, <span class="dv">17</span>, <span class="dv">98</span>, <span class="dv">35</span>, <span class="dv">11</span>, <span class="dv">13</span>)
new_list</code></pre></div>
<pre><code>## [[1]]
## [1] &quot;A&quot;
## 
## [[2]]
## [1] 1
## 
## [[3]]
## [1] 5
## 
## [[4]]
## [1] 21
## 
## [[5]]
## [1] 17
## 
## [[6]]
## [1] 98
## 
## [[7]]
## [1] 35
## 
## [[8]]
## [1] 11
## 
## [[9]]
## [1] 13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 2nd element of list
new_list[[<span class="dv">2</span>]]</code></pre></div>
<pre><code>## [1] 1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(new_list) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;Patient&quot;</span>, <span class="dv">1</span> <span class="op">:</span><span class="st"> </span><span class="kw">length</span>(new_list))
new_list</code></pre></div>
<pre><code>## $Patient1
## [1] &quot;A&quot;
## 
## $Patient2
## [1] 1
## 
## $Patient3
## [1] 5
## 
## $Patient4
## [1] 21
## 
## $Patient5
## [1] 17
## 
## $Patient6
## [1] 98
## 
## $Patient7
## [1] 35
## 
## $Patient8
## [1] 11
## 
## $Patient9
## [1] 13</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 2nd element of list
new_list[[<span class="dv">2</span>]]</code></pre></div>
<pre><code>## [1] 1</code></pre>
<ul>
<li>2D</li>
</ul>
</div>
<div id="matrix" class="section level4">
<h4><span class="header-section-number">7.3.2.6</span> matrix</h4>
<p><strong>Create Matrix</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## create numeric matrix
numeric_matrix &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">sample</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">100</span>, <span class="dt">replace =</span> <span class="ot">TRUE</span>), <span class="dt">nrow =</span> <span class="dv">10</span>, <span class="dt">ncol =</span> <span class="dv">10</span>)
<span class="kw">class</span>(numeric_matrix) ## check class</code></pre></div>
<pre><code>## [1] &quot;matrix&quot;</code></pre>
<p><strong>Check Structure</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(numeric_matrix)</code></pre></div>
<pre><code>##  int [1:10, 1:10] 3 4 6 10 3 9 10 7 7 1 ...</code></pre>
<p><strong>Get 3rd Row</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 3rd row
numeric_matrix[<span class="dv">3</span>, ]</code></pre></div>
<pre><code>##  [1] 6 7 7 5 8 5 5 4 4 7</code></pre>
<p><strong>Get 4th Column</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 4th colum
numeric_matrix[, <span class="dv">4</span>]</code></pre></div>
<pre><code>##  [1] 5 6 5 2 9 7 8 2 8 5</code></pre>
</div>
<div id="data.frame" class="section level4">
<h4><span class="header-section-number">7.3.2.7</span> data.frame</h4>
<p><strong>Get data.frame</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## built in R data.frame iris
<span class="kw">head</span>(iris)</code></pre></div>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
## 3          4.7         3.2          1.3         0.2  setosa
## 4          4.6         3.1          1.5         0.2  setosa
## 5          5.0         3.6          1.4         0.2  setosa
## 6          5.4         3.9          1.7         0.4  setosa</code></pre>
<p><strong>Check Class</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(iris)</code></pre></div>
<pre><code>## [1] &quot;data.frame&quot;</code></pre>
<p><strong>Check Structure</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(iris)</code></pre></div>
<pre><code>## &#39;data.frame&#39;:    150 obs. of  5 variables:
##  $ Sepal.Length: num  5.1 4.9 4.7 4.6 5 5.4 4.6 5 4.4 4.9 ...
##  $ Sepal.Width : num  3.5 3 3.2 3.1 3.6 3.9 3.4 3.4 2.9 3.1 ...
##  $ Petal.Length: num  1.4 1.4 1.3 1.5 1.4 1.7 1.4 1.5 1.4 1.5 ...
##  $ Petal.Width : num  0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 ...
##  $ Species     : Factor w/ 3 levels &quot;setosa&quot;,&quot;versicolor&quot;,..: 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
<p><strong>Get 3rd Row</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 3rd row
iris[<span class="dv">3</span>,]</code></pre></div>
<pre><code>##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 3          4.7         3.2          1.3         0.2  setosa</code></pre>
<p><strong>Get 4th Column</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 4th colum
iris[,<span class="dv">4</span>]</code></pre></div>
<pre><code>##   [1] 0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 0.2 0.2 0.1 0.1 0.2 0.4 0.4
##  [18] 0.3 0.3 0.3 0.2 0.4 0.2 0.5 0.2 0.2 0.4 0.2 0.2 0.2 0.2 0.4 0.1 0.2
##  [35] 0.2 0.2 0.2 0.1 0.2 0.2 0.3 0.3 0.2 0.6 0.4 0.3 0.2 0.2 0.2 0.2 1.4
##  [52] 1.5 1.5 1.3 1.5 1.3 1.6 1.0 1.3 1.4 1.0 1.5 1.0 1.4 1.3 1.4 1.5 1.0
##  [69] 1.5 1.1 1.8 1.3 1.5 1.2 1.3 1.4 1.4 1.7 1.5 1.0 1.1 1.0 1.2 1.6 1.5
##  [86] 1.6 1.5 1.3 1.3 1.3 1.2 1.4 1.2 1.0 1.3 1.2 1.3 1.3 1.1 1.3 2.5 1.9
## [103] 2.1 1.8 2.2 2.1 1.7 1.8 1.8 2.5 2.0 1.9 2.1 2.0 2.4 2.3 1.8 2.2 2.3
## [120] 1.5 2.3 2.0 2.0 1.8 2.1 1.8 1.8 1.8 2.1 1.6 1.9 2.0 2.2 1.5 1.4 2.3
## [137] 2.4 1.8 1.8 2.1 2.4 2.3 1.9 2.3 2.5 2.3 1.9 2.0 2.3 1.8</code></pre>
<p><strong>Get 3rd Row</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get 3rd row
numeric_matrix[<span class="dv">3</span>, ]</code></pre></div>
<pre><code>##  [1] 6 7 7 5 8 5 5 4 4 7</code></pre>
<p><strong>Get Species Variable</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## get variable
iris<span class="op">$</span>Species</code></pre></div>
<pre><code>##   [1] setosa     setosa     setosa     setosa     setosa     setosa    
##   [7] setosa     setosa     setosa     setosa     setosa     setosa    
##  [13] setosa     setosa     setosa     setosa     setosa     setosa    
##  [19] setosa     setosa     setosa     setosa     setosa     setosa    
##  [25] setosa     setosa     setosa     setosa     setosa     setosa    
##  [31] setosa     setosa     setosa     setosa     setosa     setosa    
##  [37] setosa     setosa     setosa     setosa     setosa     setosa    
##  [43] setosa     setosa     setosa     setosa     setosa     setosa    
##  [49] setosa     setosa     versicolor versicolor versicolor versicolor
##  [55] versicolor versicolor versicolor versicolor versicolor versicolor
##  [61] versicolor versicolor versicolor versicolor versicolor versicolor
##  [67] versicolor versicolor versicolor versicolor versicolor versicolor
##  [73] versicolor versicolor versicolor versicolor versicolor versicolor
##  [79] versicolor versicolor versicolor versicolor versicolor versicolor
##  [85] versicolor versicolor versicolor versicolor versicolor versicolor
##  [91] versicolor versicolor versicolor versicolor versicolor versicolor
##  [97] versicolor versicolor versicolor versicolor virginica  virginica 
## [103] virginica  virginica  virginica  virginica  virginica  virginica 
## [109] virginica  virginica  virginica  virginica  virginica  virginica 
## [115] virginica  virginica  virginica  virginica  virginica  virginica 
## [121] virginica  virginica  virginica  virginica  virginica  virginica 
## [127] virginica  virginica  virginica  virginica  virginica  virginica 
## [133] virginica  virginica  virginica  virginica  virginica  virginica 
## [139] virginica  virginica  virginica  virginica  virginica  virginica 
## [145] virginica  virginica  virginica  virginica  virginica  virginica 
## Levels: setosa versicolor virginica</code></pre>
</div>
</div>
<div id="detour-to-s3s4" class="section level3">
<h3><span class="header-section-number">7.3.3</span> Detour to S3/S4</h3>
<ul>
<li>S3 most of R uses</li>
<li>Bioconductor requires R packages to be written as S4 objects</li>
<li>OO field <a href="http://adv-r.had.co.nz/OO-essentials.html">guide</a></li>
<li>Closer to a typical programming language
<ul>
<li>Classes/Methods and Generics</li>
<li>Lots of Generics implemented for Bioinformatics!</li>
</ul></li>
</ul>
<p>Different way to access values. Need to use the @ symbol instead of $</p>
<ul>
<li>(@ is equivalent to $, and slot() to [[.)</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## example
object<span class="op">@</span></code></pre></div>
<div id="sparse-matrix" class="section level4">
<h4><span class="header-section-number">7.3.3.1</span> Sparse Matrix</h4>
<p>Triplet format for storing a matrix row, column, value i, p, x</p>
<p>Different from base R. Uses the S4 methods that Bioconductor uses.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sparse_matrix &lt;-<span class="st"> </span>pbmc_small<span class="op">@</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, ]
<span class="kw">class</span>(sparse_matrix)</code></pre></div>
<p><strong>ith row - 1</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sparse_matrix<span class="op">@</span>i</code></pre></div>
<p><strong>pth column - 1</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sparse_matrix<span class="op">@</span>p</code></pre></div>
<p><strong>value</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sparse_matrix<span class="op">@</span>x</code></pre></div>
<p><strong>Get First Value</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sparse_matrix[<span class="dv">2</span>,<span class="dv">1</span>]</code></pre></div>
<p><strong>dense matrix</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dense_matrix &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(sparse_matrix)
<span class="kw">class</span>(dense_matrix)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(dense_matrix)</code></pre></div>
<p><strong>Get First Value</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dense_matrix[<span class="dv">2</span>,<span class="dv">1</span>]</code></pre></div>
</div>
<div id="functions" class="section level4">
<h4><span class="header-section-number">7.3.3.2</span> Functions</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">create_function &lt;-<span class="st"> </span><span class="cf">function</span>(x, y) {
  
}</code></pre></div>
</div>
<div id="reading-files" class="section level4">
<h4><span class="header-section-number">7.3.3.3</span> Reading Files</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## read csv files
<span class="kw">read.csv</span>(<span class="st">&quot;PATH/TO/FILENAME.csv&quot;</span>)

## read tsv files
<span class="kw">read.delim</span>(<span class="st">&quot;PATH/TO/FILENAME.tsv&quot;</span>, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</code></pre></div>
</div>
</div>
</div>
<div id="more-information" class="section level2">
<h2><span class="header-section-number">7.4</span> More information</h2>
<p>You can get more information about any R commands relevant to these datatypes using by typing <code>?function</code> in an interactive session.</p>
<div id="checking-for-help-for-any-function" class="section level3">
<h3><span class="header-section-number">7.4.1</span> Checking for help for any function!</h3>
<ul>
<li>start with a ? (this indicates you need the help menu)</li>
<li>then the function name to get help on</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)

?ggplot  ## ggplot is a function, how do we use it?</code></pre></div>
</div>
</div>
<div id="grammer-of-graphics-ggplot2" class="section level2">
<h2><span class="header-section-number">7.5</span> Grammer of Graphics (ggplot2)</h2>
<div id="what-is-ggplot2" class="section level3">
<h3><span class="header-section-number">7.5.1</span> What is ggplot2?</h3>
<p>ggplot2 is an R package designed by Hadley Wickham which facilitates data plotting. In this lab, we will touch briefly on some of the features of the package. If you would like to learn more about how to use ggplot2, we would recommend reading “ggplot2 Elegant graphics for data analysis”, by Hadley Wickham or checking out his original paper on the <a href="http://vita.had.co.nz/papers/layered-grammar.pdf">package</a></p>
<ul>
<li>Data: Always start with the data, identify the dimensions you want to visualize.</li>
<li>Aesthetics: Confirm the axes based on the data dimensions, positions of various data points in the plot. Also check if any form of encoding is needed including size, shape, color and so on which are useful for plotting multiple data dimensions.</li>
<li>Scale: Do we need to scale the potential values, use a specific scale to represent multiple values or a range?</li>
<li>Geometric objects: These are popularly known as ‘geoms’. This would cover the way we would depict the data points on the visualization. Should it be points, bars, lines and so on?</li>
<li>Statistics: Do we need to show some statistical measures in the visualization like measures of central tendency, spread, confidence intervals?</li>
<li>Facets: Do we need to create subplots based on specific data dimensions?</li>
<li>Coordinate system: What kind of a coordinate system should the visualization be based on — should it be cartesian or polar?</li>
</ul>
<p><img src="images/components_grammar_graphics.png" width="90%" style="display: block; margin: auto;" /></p>
</div>
<div id="principles-of-ggplot2" class="section level3">
<h3><span class="header-section-number">7.5.2</span> Principles of ggplot2</h3>
<ul>
<li>Your data must be a dataframe if you want to plot it using ggplot2.</li>
<li>Use the <code>aes</code> mapping function to specify how variables in the dataframe map to features on your plot</li>
<li><p>Use geoms to specify how your data should be represented on your graph eg. as a scatterplot, a barplot, a boxplot etc.</p></li>
<li><p>Data: Always start with the data, identify the dimensions you want to visualize.</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Seurat)
<span class="kw">library</span>(ggplot2)

gbm &lt;-<span class="st"> </span>pbmc_small<span class="op">@</span>assays<span class="op">$</span>RNA<span class="op">@</span>data

gbm &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">as.matrix</span>(<span class="kw">t</span>(gbm)))

new_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(gbm)</code></pre></div>
<ul>
<li>Aesthetics: Confirm the axes based on the data dimensions, positions of various data points in the plot. Also check if any form of encoding is needed including size, shape, color and so on which are useful for plotting multiple data dimensions.</li>
</ul>
<p><strong>1D Plots</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new_plot_1dx &lt;-<span class="st"> </span><span class="kw">ggplot</span>(gbm, <span class="kw">aes</span>(<span class="dt">x =</span> MS4A1))
new_plot_1dx</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/1d_plot-1.png" width="90%" style="display: block; margin: auto;" /></p>
<ul>
<li><p>Scale: Do we need to scale the potential values, use a specific scale to represent multiple values or a range?</p></li>
<li><p>Geometric objects: These are popularly known as ‘geoms’. This would cover the way we would depict the data points on the visualization. Should it be points, bars, lines and so on?</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## ggplot(gbm, aes(x = MS4A1)) + geom_histogram()
## or
## new_plot_1dx &lt;- new_plot_1dx + geom_histogram()  ## reassign
new_plot_1dx <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>()</code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="2019_scWorkshop_files/figure-html/1dx_plot-1.png" width="90%" style="display: block; margin: auto;" /></p>
<div id="lab-a" class="section level4">
<h4><span class="header-section-number">7.5.2.1</span> Lab A</h4>
<p>Use different geom_ to make a different plots - try _bar() - try _density()</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new_plot_1dx <span class="op">+</span><span class="st"> </span><span class="kw">geom_density</span>()</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/dense_plot-1.png" width="90%" style="display: block; margin: auto;" /></p>
<ul>
<li>Statistics: Do we need to show some statistical measures in the visualization like measures of central tendency, spread, confidence intervals?</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(gbm, <span class="kw">aes</span>(<span class="dt">x =</span> MS4A1)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_histogram</span>() <span class="op">+</span><span class="st"> </span><span class="kw">stat_bin</span>(<span class="dt">bins =</span> <span class="dv">10</span>)</code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="2019_scWorkshop_files/figure-html/hist_plot-1.png" width="90%" style="display: block; margin: auto;" /></p>
<p><strong>2D Plots</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new_plot_2d &lt;-<span class="st"> </span><span class="kw">ggplot</span>(gbm, <span class="kw">aes</span>(<span class="dt">x =</span> MS4A1, <span class="dt">y =</span> CD79B))</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## scatter plot
new_plot_2d <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/scatter_plot-1.png" width="90%" style="display: block; margin: auto;" /></p>
</div>
<div id="lab-b" class="section level4">
<h4><span class="header-section-number">7.5.2.2</span> Lab B</h4>
<p>Use different geom_ to make a different plots - try _bar_abline() - try _bin2d()</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">new_plot_2d</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/2d_exer-1.png" width="90%" style="display: block; margin: auto;" /></p>
<ul>
<li>Adding Statisitics in 2D plots
<ul>
<li>Regression line (lm - linear model using OLS regression)</li>
</ul></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(gbm, <span class="kw">aes</span>(MS4A1, CD79B)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">stat_smooth</span>(<span class="dt">method =</span> <span class="st">&quot;lm&quot;</span>)</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/2d_exer_2-1.png" width="90%" style="display: block; margin: auto;" /></p>
<ul>
<li>Adding Text Labels</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## notice plus `+` at the end of each line, adding a new layer!
<span class="kw">ggplot</span>(gbm, <span class="kw">aes</span>(MS4A1, CD79B)) <span class="op">+</span><span class="st">            </span>## Data layer
<span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st">                            </span>## Geometry layer
<span class="st">  </span><span class="kw">stat_smooth</span>(<span class="dt">method =</span> <span class="st">&quot;lm&quot;</span>) <span class="op">+</span><span class="st">              </span>## 
<span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">rownames</span>(gbm)))</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/add_text-1.png" width="90%" style="display: block; margin: auto;" /></p>
</div>
<div id="lab-c" class="section level4">
<h4><span class="header-section-number">7.5.2.3</span> Lab C</h4>
<p>Play arund with ggplot2. See what geoms to add and layers to include.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggplot</span>(gbm, <span class="kw">aes</span>(<span class="dt">x =</span> MS4A1, <span class="dt">y =</span> CD79B))</code></pre></div>
<p><img src="2019_scWorkshop_files/figure-html/labc-1.png" width="90%" style="display: block; margin: auto;" /></p>
</div>
</div>
</div>
<div id="reference" class="section level2">
<h2><span class="header-section-number">7.6</span> Reference</h2>
<ul>
<li><a href="https://r4ds.had.co.nz/">R for Data Science</a></li>
<li><a href="http://adv-r.had.co.nz/">Advanced R</a></li>
<li><a href="http://bioconductor.org/packages/release/BiocViews.html#___Workflow">Bioconductor Workflows</a></li>
<li><a href="http://www.bioconductor.org/help/course-materials/2010/AdvancedR/S4InBioconductor.pdf">Bioconductor Presentation</a></li>
<li><a href="http://vita.had.co.nz/papers/layered-grammar.pdf">Original ggplot2 paper</a></li>
<li><a href="https://ggplot2.tidyverse.org/reference/index.html">ggplot2 reference</a></li>
<li><a href="https://github.com/rstudio/cheatsheets/blob/master/data-visualization-2.1.pdf">ggplot2 cheatsheet</a></li>
<li><a href="https://towardsdatascience.com/a-comprehensive-guide-to-the-grammar-of-graphics-for-effective-visualization-of-multi-dimensional-1f92b4ed4149">blog post</a></li>
<li><a href="https://hemberg-lab.github.io/scRNA.seq.course/introduction-to-rbioconductor.html#an-introduction-to-ggplot2">Hemberg Lab</a></li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="transcriptome-quantification.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="expression-qc-and-normalization.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["2019_scWorkshop.pdf", "2019_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
