<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>3 Understanding Sequencing Raw Data | ANALYSIS OF SINGLE CELL RNA-SEQ DATA</title>
  <meta name="description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="3 Understanding Sequencing Raw Data | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  <meta property="og:type" content="book" />
  <meta property="og:url" content="https://broadinstitute.github.io/2019_scWorkshop/" />
  <meta property="og:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />
  <meta property="og:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="github-repo" content="broadinstitute/2019_scWorkship" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Understanding Sequencing Raw Data | ANALYSIS OF SINGLE CELL RNA-SEQ DATA" />
  
  <meta name="twitter:description" content="This is a minimal example of using the bookdown package to write a book. The output format for this example is bookdown::gitbook." />
  <meta name="twitter:image" content="https://broadinstitute.github.io/2019_scWorkshop/images/cover.png" />

<meta name="author" content="Orr Ashenberg">
<meta name="author" content="Dana Silverbush">
<meta name="author" content="Kirk Gosik">


<meta name="date" content="2019-05-30">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="scrna-seq-experimental-design.html">
<link rel="next" href="data-preprocessing.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#course-overview"><i class="fa fa-check"></i><b>1.1</b> COURSE OVERVIEW</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#targeted-audience-assumed-background"><i class="fa fa-check"></i><b>1.2</b> TARGETED AUDIENCE &amp; ASSUMED BACKGROUND</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#course-format"><i class="fa fa-check"></i><b>1.3</b> COURSE FORMAT</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#getting-started"><i class="fa fa-check"></i><b>1.4</b> Getting Started</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#session-content"><i class="fa fa-check"></i><b>1.5</b> SESSION CONTENT</a><ul>
<li class="chapter" data-level="1.5.1" data-path="index.html"><a href="index.html#monday-classes-from-0930-to-1730-lunch-break-1-hr-40-min-of-total-coffee-breaks"><i class="fa fa-check"></i><b>1.5.1</b> Monday – Classes from 09:30 to 17:30 (lunch break-1 hr, 40 min of total coffee breaks)</a></li>
<li class="chapter" data-level="1.5.2" data-path="index.html"><a href="index.html#tuesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.2</b> Tuesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.3" data-path="index.html"><a href="index.html#wednesday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.3</b> Wednesday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.4" data-path="index.html"><a href="index.html#thursday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.4</b> Thursday – Classes from 09:30 to 17:30</a></li>
<li class="chapter" data-level="1.5.5" data-path="index.html"><a href="index.html#friday-classes-from-0930-to-1730"><i class="fa fa-check"></i><b>1.5.5</b> Friday – Classes from 09:30 to 17:30</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="scrna-seq-experimental-design.html"><a href="scrna-seq-experimental-design.html"><i class="fa fa-check"></i><b>2</b> scRNA-Seq Experimental Design</a></li>
<li class="chapter" data-level="3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html"><i class="fa fa-check"></i><b>3</b> Understanding Sequencing Raw Data</a><ul>
<li class="chapter" data-level="3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#class-environment"><i class="fa fa-check"></i><b>3.1</b> Class Environment</a><ul>
<li class="chapter" data-level="3.1.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#getting-into-aws-instance"><i class="fa fa-check"></i><b>3.1.1</b> Getting into AWS Instance</a></li>
</ul></li>
<li class="chapter" data-level="3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#shell-and-unix-commands"><i class="fa fa-check"></i><b>3.2</b> Shell and Unix commands</a><ul>
<li class="chapter" data-level="3.2.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#common-linux-commands"><i class="fa fa-check"></i><b>3.2.1</b> Common Linux Commands</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#file-formats"><i class="fa fa-check"></i><b>3.3</b> File formats</a><ul>
<li class="chapter" data-level="3.3.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-fastq-files"><i class="fa fa-check"></i><b>3.3.1</b> View FASTQ Files</a></li>
<li class="chapter" data-level="3.3.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#view-bam-files"><i class="fa fa-check"></i><b>3.3.2</b> View BAM Files</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#public-data-repositories"><i class="fa fa-check"></i><b>3.4</b> Public data repositories</a><ul>
<li class="chapter" data-level="3.4.1" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#cellranger10x"><i class="fa fa-check"></i><b>3.4.1</b> Cellranger/10x</a></li>
<li class="chapter" data-level="3.4.2" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#geo"><i class="fa fa-check"></i><b>3.4.2</b> GEO</a></li>
<li class="chapter" data-level="3.4.3" data-path="understanding-sequencing-raw-data.html"><a href="understanding-sequencing-raw-data.html#single-cell-portal"><i class="fa fa-check"></i><b>3.4.3</b> Single Cell Portal</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="data-preprocessing.html"><a href="data-preprocessing.html"><i class="fa fa-check"></i><b>4</b> Data Preprocessing</a></li>
<li class="chapter" data-level="5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html"><i class="fa fa-check"></i><b>5</b> Processing scRNAseq Data</a><ul>
<li class="chapter" data-level="5.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#goal"><i class="fa fa-check"></i><b>5.1</b> Goal</a></li>
<li class="chapter" data-level="5.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#further-reading"><i class="fa fa-check"></i><b>5.2</b> Further reading</a></li>
<li class="chapter" data-level="5.3" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastqc"><i class="fa fa-check"></i><b>5.3</b> FastQC</a><ul>
<li class="chapter" data-level="5.3.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#fastq-file-format"><i class="fa fa-check"></i><b>5.3.1</b> Fastq file format</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#align-the-reads"><i class="fa fa-check"></i><b>5.4</b> Align the reads</a><ul>
<li class="chapter" data-level="5.4.1" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#star-align"><i class="fa fa-check"></i><b>5.4.1</b> STAR align</a></li>
<li class="chapter" data-level="5.4.2" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#bam-file-format"><i class="fa fa-check"></i><b>5.4.2</b> Bam file format</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="processing-scrnaseq-data.html"><a href="processing-scrnaseq-data.html#visualization"><i class="fa fa-check"></i><b>5.5</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="transcriptome-quantification.html"><a href="transcriptome-quantification.html"><i class="fa fa-check"></i><b>6</b> Transcriptome Quantification</a></li>
<li class="chapter" data-level="7" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html"><i class="fa fa-check"></i><b>7</b> Introduction R/Bioconductor</a><ul>
<li class="chapter" data-level="7.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#start-environment"><i class="fa fa-check"></i><b>7.1</b> Start Environment</a></li>
<li class="chapter" data-level="7.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installing-packages"><i class="fa fa-check"></i><b>7.2</b> Installing packages</a><ul>
<li class="chapter" data-level="7.2.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#cran"><i class="fa fa-check"></i><b>7.2.1</b> CRAN</a></li>
<li class="chapter" data-level="7.2.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#github"><i class="fa fa-check"></i><b>7.2.2</b> Github</a></li>
<li class="chapter" data-level="7.2.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#bioconductor"><i class="fa fa-check"></i><b>7.2.3</b> Bioconductor</a></li>
<li class="chapter" data-level="7.2.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#source"><i class="fa fa-check"></i><b>7.2.4</b> Source</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#installation-instructions"><i class="fa fa-check"></i><b>7.3</b> Installation instructions:</a><ul>
<li class="chapter" data-level="7.3.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#classestypes"><i class="fa fa-check"></i><b>7.3.1</b> Classes/Types</a></li>
<li class="chapter" data-level="7.3.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#data-structures"><i class="fa fa-check"></i><b>7.3.2</b> Data structures</a></li>
<li class="chapter" data-level="7.3.3" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#detour-to-s3s4"><i class="fa fa-check"></i><b>7.3.3</b> Detour to S3/S4</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#more-information"><i class="fa fa-check"></i><b>7.4</b> More information</a><ul>
<li class="chapter" data-level="7.4.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#checking-for-help-for-any-function"><i class="fa fa-check"></i><b>7.4.1</b> Checking for help for any function!</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#grammer-of-graphics-ggplot2"><i class="fa fa-check"></i><b>7.5</b> Grammer of Graphics (ggplot2)</a><ul>
<li class="chapter" data-level="7.5.1" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#what-is-ggplot2"><i class="fa fa-check"></i><b>7.5.1</b> What is ggplot2?</a></li>
<li class="chapter" data-level="7.5.2" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#principles-of-ggplot2"><i class="fa fa-check"></i><b>7.5.2</b> Principles of ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="7.6" data-path="introduction-rbioconductor.html"><a href="introduction-rbioconductor.html#reference"><i class="fa fa-check"></i><b>7.6</b> Reference</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="expression-qc-and-normalization.html"><a href="expression-qc-and-normalization.html"><i class="fa fa-check"></i><b>8</b> Expression QC and Normalization</a></li>
<li class="chapter" data-level="9" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html"><i class="fa fa-check"></i><b>9</b> Data Wrangling scRNAseq</a><ul>
<li class="chapter" data-level="9.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#goal-1"><i class="fa fa-check"></i><b>9.1</b> Goal</a></li>
<li class="chapter" data-level="9.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#introduction-1"><i class="fa fa-check"></i><b>9.2</b> Introduction</a><ul>
<li class="chapter" data-level="9.2.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#load-necessary-packages"><i class="fa fa-check"></i><b>9.2.1</b> Load necessary packages</a></li>
<li class="chapter" data-level="9.2.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#read-in-nsclc-counts-matrix."><i class="fa fa-check"></i><b>9.2.2</b> Read in NSCLC counts matrix.</a></li>
<li class="chapter" data-level="9.2.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#lets-examine-the-sparse-counts-matrix"><i class="fa fa-check"></i><b>9.2.3</b> Let’s examine the sparse counts matrix</a></li>
<li class="chapter" data-level="9.2.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-big-is-the-matrix"><i class="fa fa-check"></i><b>9.2.4</b> How big is the matrix?</a></li>
<li class="chapter" data-level="9.2.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#how-much-memory-does-a-sparse-matrix-take-up-relative-to-a-dense-matrix"><i class="fa fa-check"></i><b>9.2.5</b> How much memory does a sparse matrix take up relative to a dense matrix?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#filtering-low-quality-cells"><i class="fa fa-check"></i><b>9.3</b> Filtering low-quality cells</a><ul>
<li class="chapter" data-level="9.3.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#look-at-the-summary-counts-for-genes-and-cells"><i class="fa fa-check"></i><b>9.3.1</b> Look at the summary counts for genes and cells</a></li>
<li class="chapter" data-level="9.3.2" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#plot-cells-ranked-by-their-number-of-detected-genes."><i class="fa fa-check"></i><b>9.3.2</b> Plot cells ranked by their number of detected genes.</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#beginning-with-seurat-httpsatijalab.orgseurat"><i class="fa fa-check"></i><b>9.4</b> Beginning with Seurat: <a href="http://satijalab.org/seurat/" class="uri">http://satijalab.org/seurat/</a></a><ul>
<li class="chapter" data-level="9.4.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#creating-a-seurat-object"><i class="fa fa-check"></i><b>9.4.1</b> Creating a seurat object</a></li>
</ul></li>
<li class="chapter" data-level="9.5" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-1-filter-out-low-quality-cells"><i class="fa fa-check"></i><b>9.5</b> Preprocessing step 1 : Filter out low-quality cells</a></li>
<li class="chapter" data-level="9.6" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#examine-contents-of-seurat-object"><i class="fa fa-check"></i><b>9.6</b> Examine contents of Seurat object</a><ul>
<li class="chapter" data-level="9.6.1" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#preprocessing-step-2-expression-normalization"><i class="fa fa-check"></i><b>9.6.1</b> Preprocessing step 2 : Expression normalization</a></li>
</ul></li>
<li class="chapter" data-level="9.7" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#detection-of-variable-genes-across-the-single-cells"><i class="fa fa-check"></i><b>9.7</b> Detection of variable genes across the single cells</a></li>
<li class="chapter" data-level="9.8" data-path="data-wrangling-scrnaseq.html"><a href="data-wrangling-scrnaseq.html#gene-set-expression-across-cells"><i class="fa fa-check"></i><b>9.8</b> Gene set expression across cells</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html"><i class="fa fa-check"></i><b>10</b> Identifying Cell Populations</a><ul>
<li class="chapter" data-level="10.1" data-path="identifying-cell-populations.html"><a href="identifying-cell-populations.html#google-slides"><i class="fa fa-check"></i><b>10.1</b> Google Slides</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html"><i class="fa fa-check"></i><b>11</b> Feature Selection and Cluster Analysis</a><ul>
<li class="chapter" data-level="11.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#abstract"><i class="fa fa-check"></i><b>11.1</b> Abstract</a></li>
<li class="chapter" data-level="11.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#seurat-tutorial"><i class="fa fa-check"></i><b>11.2</b> Seurat Tutorial</a><ul>
<li class="chapter" data-level="11.2.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#preprocessing-steps"><i class="fa fa-check"></i><b>11.2.1</b> Preprocessing Steps</a></li>
<li class="chapter" data-level="11.2.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#start-of-identifying-cell-types"><i class="fa fa-check"></i><b>11.2.2</b> Start of Identifying Cell Types</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#feature-selection"><i class="fa fa-check"></i><b>11.3</b> Feature Selection</a><ul>
<li class="chapter" data-level="11.3.1" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>11.3.1</b> Differential Expression Analysis</a></li>
<li class="chapter" data-level="11.3.2" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#dimensionality-reduction"><i class="fa fa-check"></i><b>11.3.2</b> Dimensionality Reduction</a></li>
<li class="chapter" data-level="11.3.3" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#independent-components-analysis-ica"><i class="fa fa-check"></i><b>11.3.3</b> Independent Components Analysis (ICA)</a></li>
<li class="chapter" data-level="11.3.4" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#clustering"><i class="fa fa-check"></i><b>11.3.4</b> Clustering</a></li>
<li class="chapter" data-level="11.3.5" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#check-clusters"><i class="fa fa-check"></i><b>11.3.5</b> Check Clusters</a></li>
<li class="chapter" data-level="11.3.6" data-path="feature-selection-and-cluster-analysis.html"><a href="feature-selection-and-cluster-analysis.html#practice-visualizingembedding"><i class="fa fa-check"></i><b>11.3.6</b> Practice Visualizing/Embedding</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="12" data-path="batch-effects.html"><a href="batch-effects.html"><i class="fa fa-check"></i><b>12</b> Batch Effects</a></li>
<li class="chapter" data-level="13" data-path="correcting-batch-effects.html"><a href="correcting-batch-effects.html"><i class="fa fa-check"></i><b>13</b> Correcting Batch Effects</a></li>
<li class="chapter" data-level="14" data-path="functional-analysis.html"><a href="functional-analysis.html"><i class="fa fa-check"></i><b>14</b> Functional Analysis</a><ul>
<li class="chapter" data-level="14.1" data-path="functional-analysis.html"><a href="functional-analysis.html#google-slides-1"><i class="fa fa-check"></i><b>14.1</b> Google Slides</a></li>
<li class="chapter" data-level="14.2" data-path="functional-analysis.html"><a href="functional-analysis.html#gene-sets-and-signatures"><i class="fa fa-check"></i><b>14.2</b> Gene sets and signatures</a><ul>
<li class="chapter" data-level="14.2.1" data-path="functional-analysis.html"><a href="functional-analysis.html#cell-cycle"><i class="fa fa-check"></i><b>14.2.1</b> Cell Cycle</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="functional-analysis.html"><a href="functional-analysis.html#pathway-analysis"><i class="fa fa-check"></i><b>14.3</b> Pathway analysis</a></li>
<li class="chapter" data-level="14.4" data-path="functional-analysis.html"><a href="functional-analysis.html#infercnv-honeybadger"><i class="fa fa-check"></i><b>14.4</b> inferCNV / honeybadger</a><ul>
<li class="chapter" data-level="14.4.1" data-path="functional-analysis.html"><a href="functional-analysis.html#create-the-infercnv-object"><i class="fa fa-check"></i><b>14.4.1</b> Create the InferCNV Object</a></li>
<li class="chapter" data-level="14.4.2" data-path="functional-analysis.html"><a href="functional-analysis.html#filtering-genes"><i class="fa fa-check"></i><b>14.4.2</b> Filtering genes</a></li>
<li class="chapter" data-level="14.4.3" data-path="functional-analysis.html"><a href="functional-analysis.html#normalize-each-cells-counts-for-sequencing-depth"><i class="fa fa-check"></i><b>14.4.3</b> Normalize each cell’s counts for sequencing depth</a></li>
<li class="chapter" data-level="14.4.4" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-anscombe-normalization"><i class="fa fa-check"></i><b>14.4.4</b> Perform Anscombe normalization</a></li>
<li class="chapter" data-level="14.4.5" data-path="functional-analysis.html"><a href="functional-analysis.html#log-transform-the-normalized-counts"><i class="fa fa-check"></i><b>14.4.5</b> Log transform the normalized counts:</a></li>
<li class="chapter" data-level="14.4.6" data-path="functional-analysis.html"><a href="functional-analysis.html#apply-maximum-bounds-to-the-expression-data-to-reduce-outlier-effects"><i class="fa fa-check"></i><b>14.4.6</b> Apply maximum bounds to the expression data to reduce outlier effects</a></li>
<li class="chapter" data-level="14.4.7" data-path="functional-analysis.html"><a href="functional-analysis.html#initial-view-before-infercnv-operations"><i class="fa fa-check"></i><b>14.4.7</b> Initial view, before inferCNV operations:</a></li>
<li class="chapter" data-level="14.4.8" data-path="functional-analysis.html"><a href="functional-analysis.html#perform-smoothing-across-chromosomes"><i class="fa fa-check"></i><b>14.4.8</b> Perform smoothing across chromosomes</a></li>
<li class="chapter" data-level="14.4.9" data-path="functional-analysis.html"><a href="functional-analysis.html#subtract-the-reference-values-from-observations-now-have-logfold-change-values"><i class="fa fa-check"></i><b>14.4.9</b> Subtract the reference values from observations, now have log(fold change) values</a></li>
<li class="chapter" data-level="14.4.10" data-path="functional-analysis.html"><a href="functional-analysis.html#invert-log-values"><i class="fa fa-check"></i><b>14.4.10</b> Invert log values</a></li>
<li class="chapter" data-level="14.4.11" data-path="functional-analysis.html"><a href="functional-analysis.html#removing-noise"><i class="fa fa-check"></i><b>14.4.11</b> Removing noise</a></li>
<li class="chapter" data-level="14.4.12" data-path="functional-analysis.html"><a href="functional-analysis.html#remove-outlier-data-points"><i class="fa fa-check"></i><b>14.4.12</b> Remove outlier data points</a></li>
<li class="chapter" data-level="14.4.13" data-path="functional-analysis.html"><a href="functional-analysis.html#find-de-genes-by-comparing-the-mutant-types-to-normal-types-basic"><i class="fa fa-check"></i><b>14.4.13</b> Find DE genes by comparing the mutant types to normal types, BASIC</a></li>
<li class="chapter" data-level="14.4.14" data-path="functional-analysis.html"><a href="functional-analysis.html#additional-information"><i class="fa fa-check"></i><b>14.4.14</b> Additional Information</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="15" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html"><i class="fa fa-check"></i><b>15</b> Pseudotime Cell Trajectories</a><ul>
<li class="chapter" data-level="15.1" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#google-slides-2"><i class="fa fa-check"></i><b>15.1</b> Google Slides</a></li>
<li class="chapter" data-level="15.2" data-path="pseudotime-cell-trajectories.html"><a href="pseudotime-cell-trajectories.html#comparison-abstract"><i class="fa fa-check"></i><b>15.2</b> Comparison Abstract</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="functional-pseudotime-analysis.html"><a href="functional-pseudotime-analysis.html"><i class="fa fa-check"></i><b>16</b> Functional Pseudotime Analysis</a></li>
<li class="chapter" data-level="17" data-path="single-cell-multiomic-technologies.html"><a href="single-cell-multiomic-technologies.html"><i class="fa fa-check"></i><b>17</b> Single Cell Multiomic Technologies</a></li>
<li class="chapter" data-level="18" data-path="cite-seq-and-scatac-seq.html"><a href="cite-seq-and-scatac-seq.html"><i class="fa fa-check"></i><b>18</b> CITE-seq and scATAC-seq</a></li>
<li class="chapter" data-level="19" data-path="single-cell-resources.html"><a href="single-cell-resources.html"><i class="fa fa-check"></i><b>19</b> Single Cell Resources</a><ul>
<li class="chapter" data-level="19.1" data-path="single-cell-resources.html"><a href="single-cell-resources.html#comprehensive-list-of-single-cell-resources"><i class="fa fa-check"></i><b>19.1</b> Comprehensive list of single-cell resources</a></li>
<li class="chapter" data-level="19.2" data-path="single-cell-resources.html"><a href="single-cell-resources.html#computational-packages-for-single-cell-analysis"><i class="fa fa-check"></i><b>19.2</b> Computational packages for single-cell analysis</a></li>
<li class="chapter" data-level="19.3" data-path="single-cell-resources.html"><a href="single-cell-resources.html#elife-commentary-on-the-human-cell-atlas"><i class="fa fa-check"></i><b>19.3</b> eLife Commentary on the Human Cell Atlas</a></li>
<li class="chapter" data-level="19.4" data-path="single-cell-resources.html"><a href="single-cell-resources.html#online-courses"><i class="fa fa-check"></i><b>19.4</b> Online courses</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">ANALYSIS OF SINGLE CELL RNA-SEQ DATA</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="understanding-sequencing-raw-data" class="section level1">
<h1><span class="header-section-number">3</span> Understanding Sequencing Raw Data</h1>
<div id="class-environment" class="section level2">
<h2><span class="header-section-number">3.1</span> Class Environment</h2>
<div id="getting-into-aws-instance" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Getting into AWS Instance</h3>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Example</span>
<span class="fu">ssh</span> -i berlin.pem ubuntu@<span class="op">&lt;</span>PUBLIC IP ADDRESS<span class="op">&gt;</span> (e.g.34.219.254.245)

<span class="co">## Actual Command</span>
<span class="fu">ssh</span> -i berlin.pem ubuntu@34.213.180.241</code></pre></div>
<!--
### Rstudio


-->
</div>
</div>
<div id="shell-and-unix-commands" class="section level2">
<h2><span class="header-section-number">3.2</span> Shell and Unix commands</h2>
<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vSavPAPY3WXe2EiqaRXJmAajFTeGcRbD1qJq4Sp4YlcF1_m7I1X89D1uC-8jDDGinqFLNO2oJNALbNx/embed?start=false&amp;loop=false&amp;delayms=3000" frameborder="0" width="760" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true">
</iframe>
<div id="common-linux-commands" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Common Linux Commands</h3>
<div id="lab-1a" class="section level4">
<h4><span class="header-section-number">3.2.1.1</span> Lab 1a</h4>
<ul>
<li>check the your present directory</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">pwd</span></code></pre></div>
<ul>
<li>check history</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">history</span></code></pre></div>
<ul>
<li>pipe history to grep to search for the cd command</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">history</span> <span class="kw">|</span> <span class="fu">grep</span> cd</code></pre></div>
<ul>
<li>put history into a history.txt file</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">history</span> <span class="op">&gt;</span> history.txt</code></pre></div>
<ul>
<li>make a directory called data</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> data</code></pre></div>
<ul>
<li>change into data directory</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">cd</span> data</code></pre></div>
<ul>
<li>move history.txt file into data directory</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mv</span> ../history.txt ./</code></pre></div>
<ul>
<li>check manual page of wget command</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">man</span> wget</code></pre></div>
<ul>
<li>redirect wget maunual page output into a file called wget.txt</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">man</span> wget <span class="op">&gt;</span> wget.txt</code></pre></div>
<ul>
<li>return the lines that contain output in the wget.txt file</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cat</span> wget.txt <span class="kw">|</span> <span class="fu">grep</span> output</code></pre></div>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">grep</span> -i output wget.txt</code></pre></div>
<ul>
<li>Compress wget.txt file</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">gzip</span> wget.txt</code></pre></div>
<ul>
<li>View Compressed file</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cat</span> wget.txt.qz</code></pre></div>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">zcat</span> wget.txt.qz</code></pre></div>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">zcat</span> wget.txt.qz <span class="kw">|</span> <span class="fu">less</span></code></pre></div>
</div>
<div id="docker-commands" class="section level4">
<h4><span class="header-section-number">3.2.1.2</span> Docker Commands</h4>
<p>Consistent compute enviornment to ensure all software that you need is on the machine and able to be used.</p>
<ul>
<li>change directory to your user directory</li>
<li>run following command to start docker</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## maybe take away the --rm so you can save the container for later</span>
<span class="co">## run from your home directory</span>
<span class="bu">cd</span> 
<span class="ex">docker</span> run --rm -it -v <span class="va">$PWD</span>/Share:/Share -v <span class="va">$PWD</span>:/mydir kdgosik/scellbern2019 bash</code></pre></div>
<p><strong>Explaination of commands</strong></p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">  <span class="ex">-</span> docker: command to run docker
  <span class="ex">-</span> run: asking docker to run a container
  <span class="ex">-</span> --rm: flag to remove the container when you exit from it
      <span class="ex">-</span> nothing will be saved from your session to access again later
      <span class="ex">-</span> this flag can be removed to keep container
  <span class="ex">-</span> -it: flag to run the container interactively
    <span class="ex">-</span> this will keep all session output displaying on the terminal
    <span class="ex">-</span> to stop container go to terminal and press Crtl+c
    <span class="ex">-v</span> <span class="va">$PWD</span>/Share:/Share: map the share directory from AWS to Share inside docker container
    <span class="ex">-v</span> <span class="va">$PWD</span>:/mydir: map your home directory to a directory inside docker container called home
  <span class="ex">-</span> kdgosik/scellbern2019: the image to run.  It will be the image into a container if not already built on your computer
    <span class="ex">-</span> [image link](https://hub.docker.com/r/kdgosik/scellbern2019)</code></pre></div>
</div>
</div>
</div>
<div id="file-formats" class="section level2">
<h2><span class="header-section-number">3.3</span> File formats</h2>
<ul>
<li>bcl</li>
<li>fastq</li>
<li>bam</li>
<li>mtx, tsv</li>
<li>hdf5 (.h5, .h5ad)</li>
</ul>
<div id="view-fastq-files" class="section level3">
<h3><span class="header-section-number">3.3.1</span> View FASTQ Files</h3>
<div id="viewing-entire-file" class="section level4">
<h4><span class="header-section-number">3.3.1.1</span> Viewing entire file</h4>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cat</span> /Share/data/Teichmann_2i_2_2_2.fastq</code></pre></div>
</div>
<div id="viewing-first-10-lines" class="section level4">
<h4><span class="header-section-number">3.3.1.2</span> Viewing first 10 lines</h4>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">head</span> /Share/data/Teichmann_2i_2_2_2.fastq</code></pre></div>
</div>
<div id="stream-viewing-with-less-command" class="section level4">
<h4><span class="header-section-number">3.3.1.3</span> Stream Viewing with less command</h4>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">less</span> /Share/data/Teichmann_2i_2_2_2.fastq</code></pre></div>
</div>
</div>
<div id="view-bam-files" class="section level3">
<h3><span class="header-section-number">3.3.2</span> View BAM Files</h3>
<div id="viewing-first-10-lines-1" class="section level4">
<h4><span class="header-section-number">3.3.2.1</span> Viewing first 10 lines</h4>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">samtools</span> view /Share/data/pbmc_1k_protein_v3_possorted_genome_bam.bam <span class="kw">|</span> <span class="fu">head</span></code></pre></div>
</div>
<div id="stream-viewing-with-less-command-1" class="section level4">
<h4><span class="header-section-number">3.3.2.2</span> Stream Viewing with less command</h4>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">samtools</span> view /Share/data/pbmc_1k_protein_v3_possorted_genome_bam.bam <span class="kw">|</span> <span class="fu">less</span></code></pre></div>
</div>
</div>
</div>
<div id="public-data-repositories" class="section level2">
<h2><span class="header-section-number">3.4</span> Public data repositories</h2>
<div id="cellranger10x" class="section level3">
<h3><span class="header-section-number">3.4.1</span> Cellranger/10x</h3>
<div id="lab-1b" class="section level4">
<h4><span class="header-section-number">3.4.1.1</span> Lab 1b</h4>
<p>10x PBMC data are hosted in <a href="https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz" class="uri">https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz</a></p>
<ul>
<li>change directory into the data directory</li>
<li>get 10x PBMC data</li>
<li>unzip data</li>
<li>explore directory</li>
<li>explore files</li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> data
<span class="fu">wget</span> https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz -O data/pbmc3k_filtered_gene_bc_matrices.tar.gz
<span class="bu">cd</span> data<span class="kw">;</span> <span class="fu">tar</span> -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz
<span class="bu">cd</span> ..</code></pre></div>
</div>
</div>
<div id="geo" class="section level3">
<h3><span class="header-section-number">3.4.2</span> GEO</h3>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81905" class="uri">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE81905</a></li>
</ul>
<div id="lab-1c" class="section level4">
<h4><span class="header-section-number">3.4.2.1</span> Lab 1c</h4>
<p><strong>Get GEO Data</strong> - <a href="ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE81nnn/GSE81905/matrix/GSE81905-GPL19057_series_matrix.txt.gz" class="uri">ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE81nnn/GSE81905/matrix/GSE81905-GPL19057_series_matrix.txt.gz</a> - <a href="ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE81nnn/GSE81905/matrix/GSE81905-GPL17021_series_matrix.txt.gz" class="uri">ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE81nnn/GSE81905/matrix/GSE81905-GPL17021_series_matrix.txt.gz</a></p>
<ul>
<li>make a directory for the files or use data directory</li>
<li>go into that directory</li>
<li>get files and place them in the directory</li>
<li>View files (try keeping in compressed format and view that way)</li>
</ul>
<p><code>bash  cd data  wget ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE81nnn/GSE81905/matrix/GSE81905-GPL19057_series_matrix.txt.gz  cd data; tar -xzf GSE81905-GPL19057_series_matrix.txt.gz  cd ..</code></p>
</div>
</div>
<div id="single-cell-portal" class="section level3">
<h3><span class="header-section-number">3.4.3</span> Single Cell Portal</h3>
<ul>
<li><a href="https://portals.broadinstitute.org/single_cell" class="uri">https://portals.broadinstitute.org/single_cell</a></li>
<li>Study: Salk Institute - Single-cell Methylome Sequencing Identifies Distinct Neuronal Populations in Mouse Frontal Cortex</li>
</ul>
<div id="lab-1d" class="section level4">
<h4><span class="header-section-number">3.4.3.1</span> Lab 1d</h4>
<ul>
<li>Get R2 fastq file from the Salk Institute study</li>
<li>Look at files</li>
</ul>
</div>
<div id="lab-1e" class="section level4">
<h4><span class="header-section-number">3.4.3.2</span> Lab 1e</h4>
<ul>
<li>Get Docker on your local computer for you to have</li>
<li>Explore Single Cell Portal</li>
<li>Explore GEO</li>
</ul>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="scrna-seq-experimental-design.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-preprocessing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": ["2019_scWorkshop.pdf", "2019_scWorkshop.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
